<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>A Framework for Consensus Partitioning • cola</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="A Framework for Consensus Partitioning">
<meta property="og:description" content="Subgroup classification is a basic task in
  genomic data analysis, especially for gene expression and DNA methylation data
  analysis. It can also be used to test the agreement to known clinical
  annotations, or to test whether there exist significant batch effects. The
  cola package provides a general framework for subgroup classification by
  consensus partitioning. It has the following features: 1. It modularizes the
  consensus partitioning processes that various methods can be easily
  integrated. 2. It provides rich visualizations for interpreting the results.
  3. It allows running multiple methods at the same time and provides
  functionalities to straightforward compare results. 4. It provides a new
  method to extract features which are more efficient to separate subgroups. 5.
  It automatically generates detailed reports for the complete analysis. 6. It allows
  applying consensus partitioning in a hierarchical manner.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">cola</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.9.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Standard consensus partitioning</li>
    <li>
      <a href="articles/cola_quick.html">A Quick Start of Using cola Package</a>
    </li>
    <li>
      <a href="articles/cola_general.html">A Framework for Consensus Partitioning</a>
    </li>
    <li>
      <a href="articles/functional_enrichment.html">Automatic Functional Enrichment on Signature genes</a>
    </li>
    <li>
      <a href="articles/predict.html">Predict Classes for New Samples</a>
    </li>
    <li>
      <a href="articles/work_with_big_datasets.html">Work with Big Datasets</a>
    </li>
    <li>
      <a href="articles/compare_partitions.html">Compare Partitions</a>
    </li>
    <li>
      <a href="articles/ATC_methods.html">ATC - More Forms</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Hierarchical partitioning</li>
    <li>
      <a href="articles/hierarchical.html">Hierarchical consensus partitioning</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Other examples</li>
    <li>
      <a href="articles/AML.html">Re-analyze an AML proteomics dataset</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">On public datasets</li>
    <li>
      <a href="articles/examples.html">Applications on public datasets</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Supplementary files</li>
    <li>
      <a href="https://jokergoo.github.io/cola_supplementary/" class="external-link">Supplementary files for cola NAR paper</a>
    </li>
    <li>
      <a href="https://cola-rh.github.io/supplementary/" class="external-link">Supplementary files for cola BIB paper</a>
    </li>
  </ul>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/jokergoo/cola/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="cola-a-general-framework-for-consensus-partitioning-">cola: A General Framework for Consensus Partitioning <img src="https://user-images.githubusercontent.com/449218/54158555-03e3af80-444b-11e9-9773-070823101263.png" width="250" align="right" style="border:4px solid black;"><a class="anchor" aria-label="anchor" href="#cola-a-general-framework-for-consensus-partitioning-"></a>
</h1></div>
<p><a href="https://github.com/jokergoo/cola/actions" class="external-link"><img src="https://github.com/jokergoo/cola/workflows/R-CMD-check/badge.svg" alt="R-CMD-check"></a> <a href="http://bioconductor.org/packages/stats/bioc/cola" class="external-link"><img src="https://bioconductor.org/shields/downloads/devel/cola.svg" alt="bioc"></a> <a href="http://bioconductor.org/checkResults/devel/bioc-LATEST/cola/" class="external-link"><img src="http://bioconductor.org//shields/lastcommit/devel/bioc/cola.svg" alt="bioc"></a></p>
<div class="section level2">
<h2 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<p>Zuguang Gu, et al., cola: an R/Bioconductor package for consensus partitioning through a general framework, Nucleic Acids Research, 2021. <a href="https://doi.org/10.1093/nar/gkaa1146" class="external-link uri">https://doi.org/10.1093/nar/gkaa1146</a></p>
<p>Zuguang Gu, et al., Improve consensus partitioning via a hierarchical procedure. Briefings in bioinformatics 2022. <a href="https://doi.org/10.1093/bib/bbac048" class="external-link uri">https://doi.org/10.1093/bib/bbac048</a></p>
</div>
<div class="section level2">
<h2 id="install">Install<a class="anchor" aria-label="anchor" href="#install"></a>
</h2>
<p><em>cola</em> is available on <a href="http://bioconductor.org/packages/devel/bioc/html/cola.html" class="external-link">Bioconductor</a>, you can install it by:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"cola"</span><span class="op">)</span></span></code></pre></div>
<p>The latest version can be installed directly from GitHub:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"jokergoo/cola"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="methods">Methods<a class="anchor" aria-label="anchor" href="#methods"></a>
</h2>
<p>The <strong>cola</strong> supports two types of consensus partitioning.</p>
<div class="section level3">
<h3 id="standard-consensus-partitioning">Standard consensus partitioning<a class="anchor" aria-label="anchor" href="#standard-consensus-partitioning"></a>
</h3>
<div class="section level4">
<h4 id="features">Features<a class="anchor" aria-label="anchor" href="#features"></a>
</h4>
<ol style="list-style-type: decimal">
<li>It modularizes the consensus clustering processes that various methods can be easily integrated in different steps of the analysis.</li>
<li>It provides rich visualizations for intepreting the results.</li>
<li>It allows running multiple methods at the same time and provides functionalities to compare results in a straightforward way.</li>
<li>It provides a new method to extract features which are more efficient to separate subgroups.</li>
<li>It generates detailed HTML reports for the complete analysis.</li>
</ol>
</div>
<div class="section level4">
<h4 id="workflow">Workflow<a class="anchor" aria-label="anchor" href="#workflow"></a>
</h4>
<p><img width="700" src="https://user-images.githubusercontent.com/449218/52628723-86af3400-2eb8-11e9-968d-b7f47a408818.png"></p>
<p>The steps of consensus partitioning is:</p>
<ol style="list-style-type: decimal">
<li>Clean the input matrix. The processing are: adjusting outliers, imputing missing values and removing rows with very small variance. This step is optional.</li>
<li>Extract subset of rows with highest scores. Here “scores” are calculated by a certain method. For gene expression analysis or methylation data analysis, <span class="math inline"><em>n</em></span> rows with highest variance are used in most cases, where the “method”, or let’s call it <strong>“the top-value method”</strong> is the variance (by <code><a href="https://rdrr.io/r/stats/cor.html" class="external-link">var()</a></code> or <code><a href="https://rdrr.io/r/stats/sd.html" class="external-link">sd()</a></code>). Note the choice of “the top-value method” can be general. It can be e.g. MAD (median absolute deviation) or any user-defined method.</li>
<li>Scale the rows in the sub-matrix (e.g. gene expression) or not (e.g. methylation data). This step is optional.</li>
<li>Randomly sample a subset of rows from the sub-matrix with probability <span class="math inline"><em>p</em></span> and perform partition on the columns of the matrix by a certain partition method, with trying different numbers of subgroups.</li>
<li>Repeat step 4 several times and collect all the partitions.</li>
<li>Perform consensus partitioning analysis and determine the best number of subgroups which gives the most stable subgrouping.</li>
<li>Apply statistical tests to find rows that show significant difference between the predicted subgroups. E.g. to extract subgroup specific genes.</li>
<li>If rows in the matrix can be associated to genes, downstream analysis such as function enrichment analysis can be performed.</li>
</ol>
</div>
<div class="section level4">
<h4 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h4>
<p>Three lines of code to perfrom <em>cola</em> analysis:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mat</span> <span class="op">=</span> <span class="fu"><a href="reference/adjust_matrix.html">adjust_matrix</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span> <span class="co"># optional</span></span>
<span><span class="va">rl</span> <span class="op">=</span> <span class="fu"><a href="reference/run_all_consensus_partition_methods.html">run_all_consensus_partition_methods</a></span><span class="op">(</span></span>
<span>    <span class="va">mat</span>, </span>
<span>    top_value_method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SD"</span>, <span class="st">"MAD"</span>, <span class="va">...</span><span class="op">)</span>,</span>
<span>    partition_method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"hclust"</span>, <span class="st">"kmeans"</span>, <span class="va">...</span><span class="op">)</span>,</span>
<span>    cores <span class="op">=</span> <span class="va">...</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/cola_report-dispatch.html">cola_report</a></span><span class="op">(</span><span class="va">rl</span>, output_dir <span class="op">=</span> <span class="va">...</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="plots">Plots<a class="anchor" aria-label="anchor" href="#plots"></a>
</h4>
<p>Following plots compare consensus heatmaps with k = 4 under all combinations of methods.</p>
<p><img src="https://user-images.githubusercontent.com/449218/52631118-3a66f280-2ebe-11e9-8dea-0172d9beab91.png"></p>
</div>
</div>
<div class="section level3">
<h3 id="hierarchical-consensus-partitioning">Hierarchical consensus partitioning<a class="anchor" aria-label="anchor" href="#hierarchical-consensus-partitioning"></a>
</h3>
<div class="section level4">
<h4 id="features-1">Features<a class="anchor" aria-label="anchor" href="#features-1"></a>
</h4>
<ol style="list-style-type: decimal">
<li>It can detect subgroups which show major differences and also moderate differences.</li>
<li>It can detect subgroups with large sizes as well as with tiny sizes.</li>
<li>It generates detailed HTML reports for the complete analysis.</li>
</ol>
</div>
<div class="section level4">
<h4 id="hierarchical-consensus-partitioning-1">Hierarchical Consensus Partitioning<a class="anchor" aria-label="anchor" href="#hierarchical-consensus-partitioning-1"></a>
</h4>
<p><img src="https://user-images.githubusercontent.com/449218/126491482-31a9496f-cc4d-4c4f-80b7-7b752d8d8d06.png" width="400"></p>
</div>
<div class="section level4">
<h4 id="usage-1">Usage<a class="anchor" aria-label="anchor" href="#usage-1"></a>
</h4>
<p>Three lines of code to perfrom hierarchical consensus partitioning analysis:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mat</span> <span class="op">=</span> <span class="fu"><a href="reference/adjust_matrix.html">adjust_matrix</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span> <span class="co"># optional</span></span>
<span><span class="va">rh</span> <span class="op">=</span> <span class="fu"><a href="reference/hierarchical_partition.html">hierarchical_partition</a></span><span class="op">(</span><span class="va">mat</span>, mc.cores <span class="op">=</span> <span class="va">...</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/cola_report-dispatch.html">cola_report</a></span><span class="op">(</span><span class="va">rh</span>, output_dir <span class="op">=</span> <span class="va">...</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="plots-1">Plots<a class="anchor" aria-label="anchor" href="#plots-1"></a>
</h4>
<p>Following figure shows the hierarchy of the subgroups.</p>
<p><img src="https://user-images.githubusercontent.com/449218/100014572-d7b2c280-2dd6-11eb-9265-a84d324122f2.png" width="300"></p>
<p>Following figure shows the signature genes.</p>
<p><img src="https://user-images.githubusercontent.com/449218/100014657-f913ae80-2dd6-11eb-9bf7-53f733e9f8f0.png" width="600"></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="license">License<a class="anchor" aria-label="anchor" href="#license"></a>
</h2>
<p>MIT @ Zuguang Gu</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://www.bioconductor.org/packages/cola" class="external-link">View on Bioconductor</a></li>
<li><a href="https://github.com/jokergoo/cola/" class="external-link">Browse source code</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing cola</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Zuguang Gu <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-7395-8709" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Zuguang Gu.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  <style>nav[data-toggle='toc'] .nav .nav {display: block;}</style>
</body>
</html>
