<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Re-analyze an AML proteomics dataset • cola</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Re-analyze an AML proteomics dataset">
<meta property="og:description" content="cola">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">cola</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.9.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Standard consensus partitioning</li>
    <li>
      <a href="../articles/cola_quick.html">A Quick Start of Using cola Package</a>
    </li>
    <li>
      <a href="../articles/cola_general.html">A Framework for Consensus Partitioning</a>
    </li>
    <li>
      <a href="../articles/functional_enrichment.html">Automatic Functional Enrichment on Signature genes</a>
    </li>
    <li>
      <a href="../articles/predict.html">Predict Classes for New Samples</a>
    </li>
    <li>
      <a href="../articles/work_with_big_datasets.html">Work with Big Datasets</a>
    </li>
    <li>
      <a href="../articles/compare_partitions.html">Compare Partitions</a>
    </li>
    <li>
      <a href="../articles/ATC_methods.html">ATC - More Forms</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Hierarchical partitioning</li>
    <li>
      <a href="../articles/hierarchical.html">Hierarchical consensus partitioning</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Other examples</li>
    <li>
      <a href="../articles/AML.html">Re-analyze an AML proteomics dataset</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">On public datasets</li>
    <li>
      <a href="../articles/examples.html">Applications on public datasets</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Supplementary files</li>
    <li>
      <a href="https://jokergoo.github.io/cola_supplementary/" class="external-link">Supplementary files for cola NAR paper</a>
    </li>
    <li>
      <a href="https://cola-rh.github.io/supplementary/" class="external-link">Supplementary files for cola BIB paper</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/jokergoo/cola/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Re-analyze an AML proteomics dataset</h1>
                        <h4 data-toc-skip class="author">Zuguang Gu (<a href="mailto:z.gu@dkfz.de" class="email">z.gu@dkfz.de</a>)</h4>
            
            <h4 data-toc-skip class="date">2024-02-27</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/jokergoo/cola/blob/HEAD/vignettes/AML.Rmd" class="external-link"><code>vignettes/AML.Rmd</code></a></small>
      <div class="hidden name"><code>AML.Rmd</code></div>

    </div>

    
    
<p>I recently came across a paper entitled <a href="https://www.sciencedirect.com/science/article/pii/S1535610822000587?via%3Dihub" class="external-link">“The proteogenomic subtypes of acute myeloid leukemia”</a> published on <em>Cancer Cell</em>, which performed subtype classifications on acute myeloid leukemia (AML) with the proteomics data. In the paper, the classification was performed by hierarchical clustering on the distance matrix of the proteomics dataset.</p>
<p>The motivations for me to reanalyze this dataset are the following two:</p>
<ol style="list-style-type: decimal">
<li>I have never analyzed proteomics data, thus I want to see how the data looks like;</li>
<li>I have never applied <a href="http://bioconductor.org/packages/release/bioc/html/cola.html" class="external-link"><strong>cola</strong></a> package on proteomics datasets. I want to see its performance and whether it can generate biological meaningful results.</li>
</ol>
<p>First the data (the Excel sheet, <code>Data S1 Discovery Cohort.xlsx</code>) is from the <a href="https://www.sciencedirect.com/science/article/pii/S1535610822000587?dgcid=coauthor#mmc7" class="external-link">Supplementary Data S1</a> of the paper.</p>
<p>I first read the clinical annotation data. There are quite a lot of columns and I don’t really need all of them.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://readxl.tidyverse.org" class="external-link">readxl</a></span><span class="op">)</span></span>
<span><span class="va">anno</span> <span class="op">=</span> <span class="fu"><a href="https://readxl.tidyverse.org/reference/read_excel.html" class="external-link">read_xlsx</a></span><span class="op">(</span><span class="st">"~/Downloads/Data S1 Discovery Cohort.xlsx"</span>, sheet <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">anno</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">anno</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/rownames-dispatch.html">rownames</a></span><span class="op">(</span><span class="va">anno</span><span class="op">)</span> <span class="op">=</span> <span class="va">anno</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span> <span class="co"># the first column is sample id</span></span>
<span><span class="va">anno</span> <span class="op">=</span> <span class="va">anno</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">anno</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    Sex Age at Diagnosis [Years] Peripheral Blasts [%] Bone marrow Blasts [%]</span></span>
<span><span class="co">## F1   f                 61.01096                     1                     40</span></span>
<span><span class="co">## F2   m                 68.27671                    56                     54</span></span>
<span><span class="co">## F4   f                 59.23836                     3                     50</span></span>
<span><span class="co">## F5   m                 65.95890                     0                     30</span></span>
<span><span class="co">## F6   f                 41.03014                    37                     51</span></span>
<span><span class="co">## F7   m                 63.93151                    25                     27</span></span>
<span><span class="co">##    Precursor                             Karyotype NPM1 FLT3 FLT3ITD VAF</span></span>
<span><span class="co">## F1      sAML 46 XX del (11)(q14q24)[15], 46,XX [5]   WT   WT          NA</span></span>
<span><span class="co">## F2   de novo                 47-48,XY,+4,+8,+13 cp   WT   WT          NA</span></span>
<span><span class="co">## F4      sAML                                46, XX   WT  ITD        4.34</span></span>
<span><span class="co">## F5   de novo                                 46,XY   WT   WT          NA</span></span>
<span><span class="co">## F6   de novo                      46,XX,-7,+8, +11   WT   WT          NA</span></span>
<span><span class="co">## F7   de novo                                 46,XY   WT   WT          NA</span></span>
<span><span class="co">##         ELN2017 FAB Induction cycles [n] Day 15 bone marrow blasts    CR1</span></span>
<span><span class="co">## F1 Intermediate  M2                    2        No blast clearance no CR1</span></span>
<span><span class="co">## F2      Adverse  M5                    2        No blast clearance    CR1</span></span>
<span><span class="co">## F4      Adverse  M4                    2           Blast clearance    CR1</span></span>
<span><span class="co">## F5      Adverse M5b                    1           Blast clearance    CR1</span></span>
<span><span class="co">## F6      Adverse  M4                    2        No blast clearance no CR1</span></span>
<span><span class="co">## F7 Intermediate  M1                    2        No blast clearance    CR1</span></span>
<span><span class="co">##    Allogeneic HSCT Death Event OS [months] Relapse Event Relapse Date</span></span>
<span><span class="co">## F1  non Transplant           1   10.316222             0         &lt;NA&gt;</span></span>
<span><span class="co">## F2  non Transplant           1    6.176591             1   2013-09-26</span></span>
<span><span class="co">## F4      Transplant           0   41.002053             0         &lt;NA&gt;</span></span>
<span><span class="co">## F5  non Transplant           1    7.227926             1   2013-08-05</span></span>
<span><span class="co">## F6      Transplant           1   55.195072             1   2016-07-20</span></span>
<span><span class="co">## F7  non Transplant           0   25.954825             1   2020-07-13</span></span>
<span><span class="co">##    RFS [months]</span></span>
<span><span class="co">## F1           NA</span></span>
<span><span class="co">## F2     2.529774</span></span>
<span><span class="co">## F4    41.002053</span></span>
<span><span class="co">## F5     5.158111</span></span>
<span><span class="co">## F6           NA</span></span>
<span><span class="co">## F7    24.574949</span></span></code></pre>
<p>Next I read the mutation data and covert it into a matrix where rows are genes and columns are samples. I also format the gene symbols that if there are mutiple genes annotated for a single mutation, only the first gene symbol is used.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mut</span> <span class="op">=</span> <span class="fu"><a href="https://readxl.tidyverse.org/reference/read_excel.html" class="external-link">read_xlsx</a></span><span class="op">(</span><span class="st">"~/Downloads/Data S1 Discovery Cohort.xlsx"</span>, sheet <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="va">mut</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">mut</span><span class="op">)</span></span>
<span><span class="va">mut</span><span class="op">[</span>, <span class="st">"Gene"</span><span class="op">]</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">",.*$"</span>, <span class="st">""</span>, <span class="va">mut</span><span class="op">[</span>, <span class="st">"Gene"</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/hadley/reshape" class="external-link">reshape2</a></span><span class="op">)</span></span>
<span><span class="va">f</span> <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>, collapse <span class="op">=</span> <span class="st">";"</span><span class="op">)</span></span>
<span><span class="va">mut</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/reshape2/man/cast.html" class="external-link">acast</a></span><span class="op">(</span><span class="va">mut</span>, <span class="va">Gene</span> <span class="op">~</span> <span class="va">Pat_ID</span>, <span class="va">f</span>, value.var <span class="op">=</span> <span class="st">"Type"</span><span class="op">)</span></span>
<span><span class="va">mut</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##         F1    F10 F101  F102 F103 </span></span>
<span><span class="co">## ANKRD26 ""    ""  ""    ""   ""   </span></span>
<span><span class="co">## ASXL1   ""    ""  ""    ""   ""   </span></span>
<span><span class="co">## ASXL2   ""    ""  ""    ""   ""   </span></span>
<span><span class="co">## ATM     ""    ""  "SNV" ""   ""   </span></span>
<span><span class="co">## ATRX    "SNV" ""  ""    ""   "SNV"</span></span></code></pre>
<p>Next I read the proteomics data matrix. <strong>Note here I use the matrix where missing values are all imputed.</strong> In the proteomics data matrix, rows actually correspond to proteins. For the ease of discussion, I always write “genes” to describe rows of the proteomics data matrix.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mat</span> <span class="op">=</span> <span class="fu"><a href="https://readxl.tidyverse.org/reference/read_excel.html" class="external-link">read_xlsx</a></span><span class="op">(</span><span class="st">"~/Downloads/Data S1 Discovery Cohort.xlsx"</span>, sheet <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span>
<span><span class="va">mat</span> <span class="op">=</span> <span class="va">mat</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span>, <span class="op">]</span></span>
<span><span class="va">gene</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unname.html" class="external-link">unname</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">mat</span><span class="op">[</span>, <span class="st">"PG.Genes"</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">mat</span> <span class="op">=</span> <span class="va">mat</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">177</span><span class="op">]</span></span>
<span><span class="va">cn</span> <span class="op">=</span> <span class="fu"><a href="../reference/colnames-dispatch.html">colnames</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span></span>
<span><span class="va">mat</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">gene</span><span class="op">)</span>, <span class="fl">177</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/colnames-dispatch.html">colnames</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span> <span class="op">=</span> <span class="va">cn</span></span>
<span></span>
<span><span class="va">l</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">gene</span><span class="op">)</span></span>
<span><span class="va">gene</span> <span class="op">=</span> <span class="va">gene</span><span class="op">[</span><span class="op">!</span><span class="va">l</span><span class="op">]</span></span>
<span><span class="va">mat</span> <span class="op">=</span> <span class="va">mat</span><span class="op">[</span><span class="op">!</span><span class="va">l</span>, <span class="op">]</span></span></code></pre></div>
<p>Similar as the mutation matrix, I also performed cleaning on gene symbols.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"^;"</span>, <span class="st">""</span>, <span class="va">gene</span><span class="op">)</span></span>
<span><span class="va">gene</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">";.*$"</span>, <span class="st">""</span>, <span class="va">gene</span><span class="op">)</span></span></code></pre></div>
<p>There are some genes wrongly formatted to dates in Excel. Here I convert them back;</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"^\\d+-"</span>, <span class="va">gene</span>, value <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "5-Mar"  "15-Sep" "6-Sep"  "2-Sep"  "7-Sep"  "1-Sep"  "8-Sep"  "5-Sep" </span></span>
<span><span class="co">##  [9] "11-Sep" "9-Sep"</span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"^(\\d+)-Mar$"</span>, <span class="st">"MAR\\1"</span>, <span class="va">gene</span><span class="op">)</span></span>
<span><span class="va">gene</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"^(\\d+)-Sep$"</span>, <span class="st">"SEP\\1"</span>, <span class="va">gene</span><span class="op">)</span></span></code></pre></div>
<p>Now I can check whether all genes are unique:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/any.html" class="external-link">any</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/duplicated.html" class="external-link">duplicated</a></span><span class="op">(</span><span class="va">gene</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] FALSE</span></span></code></pre>
<p>Since now all genes are unique, I can assign them as row names of the matrix.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/rownames-dispatch.html">rownames</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span> <span class="op">=</span> <span class="va">gene</span></span></code></pre></div>
<p>Next I adjust the order of the annotation table and mutation matrix to make them consistent to the proteomics data matrix.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anno</span> <span class="op">=</span> <span class="va">anno</span><span class="op">[</span><span class="fu"><a href="../reference/colnames-dispatch.html">colnames</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">mut2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="st">""</span>, nrow <span class="op">=</span> <span class="fu"><a href="../reference/nrow-dispatch.html">nrow</a></span><span class="op">(</span><span class="va">mut</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fu"><a href="../reference/ncol-dispatch.html">ncol</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/rownames-dispatch.html">rownames</a></span><span class="op">(</span><span class="va">mut2</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="../reference/rownames-dispatch.html">rownames</a></span><span class="op">(</span><span class="va">mut</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/colnames-dispatch.html">colnames</a></span><span class="op">(</span><span class="va">mut2</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="../reference/colnames-dispatch.html">colnames</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span></span>
<span><span class="va">mut2</span><span class="op">[</span><span class="fu"><a href="../reference/rownames-dispatch.html">rownames</a></span><span class="op">(</span><span class="va">mut</span><span class="op">)</span>, <span class="fu"><a href="../reference/colnames-dispatch.html">colnames</a></span><span class="op">(</span><span class="va">mut</span><span class="op">)</span><span class="op">]</span> <span class="op">=</span> <span class="va">mut</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">anno</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 177  20</span></span></code></pre>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">mut2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1]  79 177</span></span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 5193  177</span></span></code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/identical.html" class="external-link">identical</a></span><span class="op">(</span><span class="fu"><a href="../reference/rownames-dispatch.html">rownames</a></span><span class="op">(</span><span class="va">anno</span><span class="op">)</span>, <span class="fu"><a href="../reference/colnames-dispatch.html">colnames</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/identical.html" class="external-link">identical</a></span><span class="op">(</span><span class="fu"><a href="../reference/colnames-dispatch.html">colnames</a></span><span class="op">(</span><span class="va">mut2</span><span class="op">)</span>, <span class="fu"><a href="../reference/colnames-dispatch.html">colnames</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<p>In the paper, the subtype classification was based on hierarchical clustering on the distance matrix of samples. Here I perform the subtype classification by consensus partitioning, with the <strong>cola</strong> package.</p>
<p>In the following code, <em>ATC</em> is used to extract top features and <em>skmeans</em> is used for partitioning.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/cola" class="external-link">cola</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">12345</span><span class="op">)</span></span>
<span><span class="va">mat</span> <span class="op">=</span> <span class="fu"><a href="../reference/adjust_matrix.html">adjust_matrix</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span> <span class="co"># simply to remove outliers if there is any</span></span>
<span><span class="va">res</span> <span class="op">=</span> <span class="fu"><a href="../reference/consensus_partition.html">consensus_partition</a></span><span class="op">(</span><span class="va">mat</span>, top_value_method <span class="op">=</span> <span class="st">"ATC"</span>, partition_method <span class="op">=</span> <span class="st">"skmeans"</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><code><a href="../reference/collect_plots-dispatch.html">collect_plots()</a></code> generates a lot of plots that help to decide which number of groups gives “the best partition”.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/collect_plots-dispatch.html">collect_plots</a></span><span class="op">(</span><span class="va">res</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="AML_files/figure-html/unnamed-chunk-11-1.png" width="768" style="display: block; margin: auto;"></p>
<p><code><a href="../reference/collect_stats-dispatch.html">collect_stats()</a></code> applies several quantitative measures for deciding the best k.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/collect_stats-dispatch.html">collect_stats</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<p><img src="AML_files/figure-html/unnamed-chunk-12-1.png" width="768" style="display: block; margin: auto;"></p>
<p>All the plots show 4 is the optimized number of subgroups.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">best_k</span> <span class="op">=</span> <span class="fl">4</span></span></code></pre></div>
<p>The subtype labels are saved in the variable <code>cl</code>. And the size of each subgroup are:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cl</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_classes-dispatch.html">get_classes</a></span><span class="op">(</span><span class="va">res</span>, k <span class="op">=</span> <span class="va">best_k</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">cl</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## cl</span></span>
<span><span class="co">##  1  2  3  4 </span></span>
<span><span class="co">## 69 49 33 26</span></span></code></pre>
<p>Next I make the signature heatmap which contains genes showing significantly difference between subgroups. The object <code>tb</code> contains the results of the signature analysis.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tb</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_signatures-dispatch.html">get_signatures</a></span><span class="op">(</span><span class="va">res</span>, k <span class="op">=</span> <span class="va">best_k</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="AML_files/figure-html/unnamed-chunk-15-1.png" width="672" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">tb</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   which_row          fdr      p_value   mean_1   mean_2   mean_3   mean_4</span></span>
<span><span class="co">## 1         1 1.314847e-13 1.245953e-14 15.74037 16.34356 16.05322 16.32805</span></span>
<span><span class="co">## 2         2 1.957486e-04 1.140445e-04 13.48803 13.87960 13.38700 14.05535</span></span>
<span><span class="co">## 3         3 9.774668e-13 1.113595e-13 15.42199 16.03708 15.30332 16.49124</span></span>
<span><span class="co">## 4         4 8.892207e-08 2.646211e-08 14.33122 15.03357 15.10674 14.29921</span></span>
<span><span class="co">## 5         5 1.412425e-11 2.061517e-12 15.60644 16.27084 15.95040 16.32491</span></span>
<span><span class="co">## 6         6 3.339018e-02 2.821209e-02 18.36344 19.01163 18.20078 18.71306</span></span>
<span><span class="co">##   group_diff scaled_mean_1 scaled_mean_2 scaled_mean_3 scaled_mean_4</span></span>
<span><span class="co">## 1  0.6031946    -0.6620782     0.6121620  -0.001177959     0.5793944</span></span>
<span><span class="co">## 2  0.6683547    -0.2350354     0.2942459  -0.371595263     0.5318077</span></span>
<span><span class="co">## 3  1.1879209    -0.3971493     0.3986856  -0.550696614     0.9862904</span></span>
<span><span class="co">## 4  0.8075347    -0.4030978     0.4389542   0.526675741    -0.4414770</span></span>
<span><span class="co">## 5  0.7184693    -0.6126957     0.5320413  -0.020055290     0.6251964</span></span>
<span><span class="co">## 6  0.8108483    -0.1444529     0.3199708  -0.261002551     0.1060503</span></span>
<span><span class="co">##   group_diff_scaled km</span></span>
<span><span class="co">## 1         1.2742402  3</span></span>
<span><span class="co">## 2         0.9034030  3</span></span>
<span><span class="co">## 3         1.5369870  3</span></span>
<span><span class="co">## 4         0.9681528  2</span></span>
<span><span class="co">## 5         1.2378921  3</span></span>
<span><span class="co">## 6         0.5809734  3</span></span></code></pre>
<p>Next I make an integrative visualization that contains subgroup classification, annotations, signatures and mutations. Note in the following plot, columns always correspond to the same samples in the heatmap and in the oncoprint.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m_sig</span> <span class="op">=</span> <span class="va">mat</span><span class="op">[</span><span class="va">tb</span><span class="op">$</span><span class="va">which_row</span>, <span class="op">]</span></span>
<span><span class="va">m_sig</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/scale.html" class="external-link">scale</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">m_sig</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/ComplexHeatmap" class="external-link">ComplexHeatmap</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/circlize" class="external-link">circlize</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></span><span class="op">(</span><span class="va">m_sig</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html" class="external-link">colorRamp2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">2</span>, <span class="fl">0</span>, <span class="fl">2</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"green"</span>, <span class="st">"white"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    top_annotation <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/HeatmapAnnotation.html" class="external-link">HeatmapAnnotation</a></span><span class="op">(</span>df <span class="op">=</span> <span class="va">anno</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"NPM1"</span>, <span class="st">"FLT3"</span>, <span class="st">"ELN2017"</span>, <span class="st">"FAB"</span><span class="op">)</span><span class="op">]</span>,</span>
<span>        col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>NPM1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"WT"</span> <span class="op">=</span> <span class="st">"#EEEEEE"</span>, <span class="st">"Mut"</span> <span class="op">=</span> <span class="st">"pink"</span><span class="op">)</span>,</span>
<span>                   FLT3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"WT"</span> <span class="op">=</span> <span class="st">"#EEEEEE"</span>, <span class="st">"ITD"</span> <span class="op">=</span> <span class="st">"brown"</span>, <span class="st">"TKD"</span> <span class="op">=</span> <span class="st">"green"</span><span class="op">)</span>,</span>
<span>                   ELN2017 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Intermediate"</span> <span class="op">=</span> <span class="st">"yellow"</span>, <span class="st">"Favorable"</span> <span class="op">=</span> <span class="st">"blue"</span>, <span class="st">"Adverse"</span> <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">)</span>,</span>
<span>    row_split <span class="op">=</span> <span class="va">tb</span><span class="op">$</span><span class="va">km</span>, column_split <span class="op">=</span> <span class="va">cl</span>, show_row_names <span class="op">=</span> <span class="cn">FALSE</span>, show_column_names <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    cluster_row_slices <span class="op">=</span> <span class="cn">FALSE</span>, cluster_column_slices <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    show_row_dend <span class="op">=</span> <span class="cn">FALSE</span>, show_column_dend <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/pct_v_pct.html" class="external-link">%v%</a></span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/oncoPrint.html" class="external-link">oncoPrint</a></span><span class="op">(</span><span class="va">mut2</span>, alter_fun <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"background"</span> <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">w</span>, <span class="va">h</span><span class="op">)</span> </span>
<span>        <span class="fu"><a href="https://rdrr.io/r/grid/grid.rect.html" class="external-link">grid.rect</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">w</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">2</span>, <span class="st">"pt"</span><span class="op">)</span>, <span class="va">h</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">2</span>, <span class="st">"pt"</span><span class="op">)</span>, gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"#CCCCCC"</span>, col <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    <span class="st">"Deletion"</span> <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">w</span>, <span class="va">h</span><span class="op">)</span> </span>
<span>        <span class="fu"><a href="https://rdrr.io/r/grid/grid.rect.html" class="external-link">grid.rect</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">w</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">2</span>, <span class="st">"pt"</span><span class="op">)</span>, <span class="va">h</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">2</span>, <span class="st">"pt"</span><span class="op">)</span>, gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"orange"</span>, col <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    <span class="st">"Insertion"</span> <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">w</span>, <span class="va">h</span><span class="op">)</span> </span>
<span>        <span class="fu"><a href="https://rdrr.io/r/grid/grid.rect.html" class="external-link">grid.rect</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">w</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">2</span>, <span class="st">"pt"</span><span class="op">)</span>, <span class="va">h</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">2</span>, <span class="st">"pt"</span><span class="op">)</span>, gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"red"</span>, col <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    <span class="st">"MNV"</span> <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">w</span>, <span class="va">h</span><span class="op">)</span> </span>
<span>        <span class="fu"><a href="https://rdrr.io/r/grid/grid.rect.html" class="external-link">grid.rect</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">w</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">2</span>, <span class="st">"pt"</span><span class="op">)</span>, <span class="va">h</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">2</span>, <span class="st">"pt"</span><span class="op">)</span>, gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"darkgreen"</span>, col <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    <span class="st">"Replacement"</span> <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">w</span>, <span class="va">h</span><span class="op">)</span> </span>
<span>        <span class="fu"><a href="https://rdrr.io/r/grid/grid.rect.html" class="external-link">grid.rect</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">w</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">2</span>, <span class="st">"pt"</span><span class="op">)</span>, <span class="va">h</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">2</span>, <span class="st">"pt"</span><span class="op">)</span>, gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"pink"</span>, col <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    <span class="st">"SNV"</span> <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">w</span>, <span class="va">h</span><span class="op">)</span> </span>
<span>        <span class="fu"><a href="https://rdrr.io/r/grid/grid.rect.html" class="external-link">grid.rect</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">w</span>  <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">2</span>, <span class="st">"pt"</span><span class="op">)</span>, <span class="va">h</span><span class="op">*</span><span class="fl">0.33</span>, gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"blue"</span>, col <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">)</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Deletion"</span> <span class="op">=</span> <span class="st">"orange"</span>, <span class="st">"Insertion"</span> <span class="op">=</span> <span class="st">"red"</span>, <span class="st">"MNV"</span> <span class="op">=</span> <span class="st">"darkgreen"</span>, <span class="st">"Replacement"</span> <span class="op">=</span> <span class="st">"pink"</span>, <span class="st">"SNV"</span> <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span>,</span>
<span>    height <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">24</span>, <span class="st">"cm"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="AML_files/figure-html/unnamed-chunk-16-1.png" width="100%" style="display: block; margin: auto;"></p>
<p>I use <code><a href="../reference/test_between_factors.html">test_between_factors()</a></code> to check the correlation between cola subgroups and groups in the annotation table <code>anno</code>. If the annotation values are numeric, one-way ANOVA is used and if the annotation values are categorical, chi-square test is used. In the following result, values are p-values from the corresponding tests.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anno2</span> <span class="op">=</span> <span class="va">anno</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Sex"</span>,</span>
<span>                 <span class="st">"Age at Diagnosis [Years]"</span>,</span>
<span>                 <span class="st">"Peripheral Blasts [%]"</span>,</span>
<span>                 <span class="st">"Bone marrow Blasts [%]"</span>,</span>
<span>                 <span class="st">"Precursor"</span>,</span>
<span>                 <span class="st">"NPM1"</span>,</span>
<span>                 <span class="st">"FLT3"</span>,</span>
<span>                 <span class="st">"ELN2017"</span>,</span>
<span>                 <span class="st">"Day 15 bone marrow blasts"</span>,</span>
<span>                 <span class="st">"CR1"</span>,</span>
<span>                 <span class="st">"Allogeneic HSCT"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/test_between_factors.html">test_between_factors</a></span><span class="op">(</span><span class="va">anno2</span>, <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">cl</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                           as.character(cl)</span></span>
<span><span class="co">## Sex                           0.5880710934</span></span>
<span><span class="co">## Age at Diagnosis [Years]      0.9416598486</span></span>
<span><span class="co">## Peripheral Blasts [%]         0.0001197542</span></span>
<span><span class="co">## Bone marrow Blasts [%]        0.0007065670</span></span>
<span><span class="co">## Precursor                     0.0418247843</span></span>
<span><span class="co">## NPM1                          0.5251085800</span></span>
<span><span class="co">## FLT3                          0.2800890015</span></span>
<span><span class="co">## ELN2017                       0.1660291677</span></span>
<span><span class="co">## Day 15 bone marrow blasts     0.5635794095</span></span>
<span><span class="co">## CR1                           0.1001524214</span></span>
<span><span class="co">## Allogeneic HSCT               0.0017604648</span></span></code></pre>
<p>The tests show annotations of “Peripheral Blasts [%]”, “Bone marrow Blasts [%]” and “Allogeneic HSCT” have significant relations to the cola subgroup classification.</p>
<p>Next chunk of code visualizes how the different annotations distribute in the cola subgroups. Boxplots are used when the annotation values are numeric and barplots are used when the annotation values are categorical.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="../reference/ncol-dispatch.html">ncol</a></span><span class="op">(</span><span class="va">anno2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">is.numeric</a></span><span class="op">(</span><span class="va">anno2</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html" class="external-link">boxplot</a></span><span class="op">(</span><span class="va">anno2</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">~</span> <span class="va">cl</span>, xlab <span class="op">=</span> <span class="st">"cola group"</span>, ylab <span class="op">=</span> <span class="fu"><a href="../reference/colnames-dispatch.html">colnames</a></span><span class="op">(</span><span class="va">anno2</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, main <span class="op">=</span> <span class="fu"><a href="../reference/colnames-dispatch.html">colnames</a></span><span class="op">(</span><span class="va">anno2</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>        <span class="va">tbv</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">anno2</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, <span class="va">cl</span><span class="op">)</span></span>
<span>        <span class="va">level</span> <span class="op">=</span> <span class="fu"><a href="../reference/rownames-dispatch.html">rownames</a></span><span class="op">(</span><span class="va">tbv</span><span class="op">)</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html" class="external-link">barplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">anno2</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, <span class="va">cl</span><span class="op">)</span>, xlab <span class="op">=</span> <span class="st">"cola group"</span>, ylab <span class="op">=</span> <span class="fu"><a href="../reference/colnames-dispatch.html">colnames</a></span><span class="op">(</span><span class="va">anno2</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>,</span>
<span>            col <span class="op">=</span> <span class="fl">1</span><span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">level</span><span class="op">)</span>, main <span class="op">=</span> <span class="fu"><a href="../reference/colnames-dispatch.html">colnames</a></span><span class="op">(</span><span class="va">anno2</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="st">"topright"</span>, pch <span class="op">=</span> <span class="fl">15</span>, legend <span class="op">=</span> <span class="va">level</span>, col <span class="op">=</span> <span class="fl">1</span><span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">level</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="AML_files/figure-html/unnamed-chunk-18-1.png" width="960" style="display: block; margin: auto;"></p>
<p>In the annotation table, there is also the survival data, thus, I can check the survival curves in the four cola subgroups.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anno</span><span class="op">$</span><span class="va">cola</span> <span class="op">=</span> <span class="va">cl</span>  <span class="co"># append `cl` to the annotation table</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/therneau/survival" class="external-link">"survival"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://rpkgs.datanovia.com/survminer/index.html" class="external-link">"survminer"</a></span><span class="op">)</span></span>
<span><span class="va">fit</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/survival/man/survfit.html" class="external-link">survfit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/survival/man/Surv.html" class="external-link">Surv</a></span><span class="op">(</span><span class="va">`OS [months]`</span>, <span class="va">`Death Event`</span><span class="op">)</span> <span class="op">~</span> <span class="va">cola</span>, data <span class="op">=</span> <span class="va">anno</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/survminer/man/ggsurvplot.html" class="external-link">ggsurvplot</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span></code></pre></div>
<p><img src="AML_files/figure-html/unnamed-chunk-19-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Next I check the pairwise survival difference. Values in the results are p-values. We can see cola group 1 and group 4 show the most significant survival difference.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/survminer/man/pairwise_survdiff.html" class="external-link">pairwise_survdiff</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/survival/man/Surv.html" class="external-link">Surv</a></span><span class="op">(</span><span class="va">`OS [months]`</span>, <span class="va">`Death Event`</span><span class="op">)</span> <span class="op">~</span> <span class="va">cola</span>, data <span class="op">=</span> <span class="va">anno</span>, p.adjust.method <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">##  Pairwise comparisons using Log-Rank test </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## data:  anno and cola </span></span>
<span><span class="co">## </span></span>
<span><span class="co">##   1     2     3    </span></span>
<span><span class="co">## 2 0.074 -     -    </span></span>
<span><span class="co">## 3 0.624 0.346 -    </span></span>
<span><span class="co">## 4 0.026 0.565 0.150</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## P value adjustment method: none</span></span></code></pre>
<p>Recall the following signature heatmap which I showed before:</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/get_signatures-dispatch.html">get_signatures</a></span><span class="op">(</span><span class="va">res</span>, k <span class="op">=</span> <span class="va">best_k</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="AML_files/figure-html/unnamed-chunk-21-1.png" width="672" style="display: block; margin: auto;"></p>
<p>In rows, there are five row groups. I apply Gene Ontology enrichment on the genes in the five row groups separately, with the function <code><a href="../reference/functional_enrichment-dispatch.html">functional_enrichment()</a></code>. The enrichment results are visualized by the package <strong>simplifyEnrichment</strong>.</p>
<p>In the following heatmap, the left heatmap (green-red) illustrates the adjusted p-values (FDRs) of GO terms enriched in the corresponding group of genes. The right heatmap shows the similarity of GO terms. The word clouds illustrates the overrepresented keywords from the GO terms in each GO cluster.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lt</span> <span class="op">=</span> <span class="fu"><a href="../reference/functional_enrichment-dispatch.html">functional_enrichment</a></span><span class="op">(</span><span class="va">res</span>, k <span class="op">=</span> <span class="va">best_k</span>, ont <span class="op">=</span> <span class="st">"BP"</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/simplifyEnrichment" class="external-link">simplifyEnrichment</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/simplifyEnrichment/man/simplifyGOFromMultipleLists.html" class="external-link">simplifyGOFromMultipleLists</a></span><span class="op">(</span><span class="va">lt</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="AML_files/figure-html/unnamed-chunk-22-1.png" width="864" style="display: block; margin: auto;"></p>
<p>I also apply Gene Ontology enrichment with the Cellular Component (CC) terms. Similar code as below:</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lt</span> <span class="op">=</span> <span class="fu"><a href="../reference/functional_enrichment-dispatch.html">functional_enrichment</a></span><span class="op">(</span><span class="va">res</span>, k <span class="op">=</span> <span class="va">best_k</span>, ont <span class="op">=</span> <span class="st">"CC"</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/simplifyEnrichment" class="external-link">simplifyEnrichment</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/simplifyEnrichment/man/simplifyGOFromMultipleLists.html" class="external-link">simplifyGOFromMultipleLists</a></span><span class="op">(</span><span class="va">lt</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="AML_files/figure-html/unnamed-chunk-23-1.png" width="864" style="display: block; margin: auto;"></p>
<p>In the original paper, all 177 samples were classified into six subgroups. However, I cannot find the subgroup classification results in the paper or in the supplementary files, thus I cannot compare cola classification to the classification performed in the paper.</p>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">## Platform: x86_64-apple-darwin20 (64-bit)</span></span>
<span><span class="co">## Running under: macOS Ventura 13.2.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Random number generation:</span></span>
<span><span class="co">##  RNG:     L'Ecuyer-CMRG </span></span>
<span><span class="co">##  Normal:  Inversion </span></span>
<span><span class="co">##  Sample:  Rejection </span></span>
<span><span class="co">##  </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] C/UTF-8/C/C/C/C</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Europe/Berlin</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] grid      stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] simplifyEnrichment_1.11.1 BiocGenerics_0.48.1      </span></span>
<span><span class="co">##  [3] survminer_0.4.9           ggpubr_0.6.0             </span></span>
<span><span class="co">##  [5] ggplot2_3.4.4             survival_3.5-8           </span></span>
<span><span class="co">##  [7] circlize_0.4.15           ComplexHeatmap_2.18.0    </span></span>
<span><span class="co">##  [9] doRNG_1.8.6               rngtools_1.5.2           </span></span>
<span><span class="co">## [11] foreach_1.5.2             cola_2.9.1               </span></span>
<span><span class="co">## [13] reshape2_1.4.4            readxl_1.4.3             </span></span>
<span><span class="co">## [15] knitr_1.44               </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] splines_4.3.1           bitops_1.0-7            ggplotify_0.1.2        </span></span>
<span><span class="co">##   [4] tibble_3.2.1            cellranger_1.1.0        polyclip_1.10-6        </span></span>
<span><span class="co">##   [7] XML_3.99-0.14           lifecycle_1.0.4         rstatix_0.7.2          </span></span>
<span><span class="co">##  [10] doParallel_1.0.17       rprojroot_2.0.3         NLP_0.2-1              </span></span>
<span><span class="co">##  [13] lattice_0.21-9          MASS_7.3-60             backports_1.4.1        </span></span>
<span><span class="co">##  [16] magrittr_2.0.3          sass_0.4.8              rmarkdown_2.25         </span></span>
<span><span class="co">##  [19] jquerylib_0.1.4         yaml_2.3.7              cowplot_1.1.1          </span></span>
<span><span class="co">##  [22] DBI_1.1.3               RColorBrewer_1.1-3      eulerr_7.0.0           </span></span>
<span><span class="co">##  [25] abind_1.4-5             zlibbioc_1.46.0         purrr_1.0.2            </span></span>
<span><span class="co">##  [28] ggraph_2.1.0            RCurl_1.98-1.12         yulab.utils_0.1.0      </span></span>
<span><span class="co">##  [31] tweenr_2.0.2            GenomeInfoDbData_1.2.10 IRanges_2.36.0         </span></span>
<span><span class="co">##  [34] KMsurv_0.1-5            S4Vectors_0.40.2        enrichplot_1.20.3      </span></span>
<span><span class="co">##  [37] tm_0.7-11               ggrepel_0.9.4           irlba_2.3.5.1          </span></span>
<span><span class="co">##  [40] tidytree_0.4.5          genefilter_1.82.1       annotate_1.78.0        </span></span>
<span><span class="co">##  [43] brew_1.0-8              commonmark_1.9.0        pkgdown_2.0.7          </span></span>
<span><span class="co">##  [46] codetools_0.2-19        DOSE_3.26.2             xml2_1.3.6             </span></span>
<span><span class="co">##  [49] ggforce_0.4.1           tidyselect_1.2.0        shape_1.4.6            </span></span>
<span><span class="co">##  [52] aplot_0.2.2             farver_2.1.1            viridis_0.6.4          </span></span>
<span><span class="co">##  [55] matrixStats_1.2.0       stats4_4.3.1            jsonlite_1.8.8         </span></span>
<span><span class="co">##  [58] GetoptLong_1.0.5        tidygraph_1.2.3         iterators_1.0.14       </span></span>
<span><span class="co">##  [61] systemfonts_1.0.5       tools_4.3.1             treeio_1.24.3          </span></span>
<span><span class="co">##  [64] ragg_1.2.6              Rcpp_1.0.11             glue_1.6.2             </span></span>
<span><span class="co">##  [67] gridExtra_2.3           xfun_0.40               qvalue_2.32.0          </span></span>
<span><span class="co">##  [70] MatrixGenerics_1.12.3   GenomeInfoDb_1.36.4     dplyr_1.1.3            </span></span>
<span><span class="co">##  [73] withr_2.5.2             fastmap_1.1.1           fansi_1.0.5            </span></span>
<span><span class="co">##  [76] digest_0.6.33           R6_2.5.1                gridGraphics_0.5-1     </span></span>
<span><span class="co">##  [79] microbenchmark_1.4.10   textshaping_0.3.7       colorspace_2.1-0       </span></span>
<span><span class="co">##  [82] GO.db_3.17.0            Cairo_1.6-2             markdown_1.10          </span></span>
<span><span class="co">##  [85] RSQLite_2.3.1           utf8_1.2.3              tidyr_1.3.0            </span></span>
<span><span class="co">##  [88] generics_0.1.3          data.table_1.14.8       graphlayouts_1.0.1     </span></span>
<span><span class="co">##  [91] httr_1.4.7              scatterpie_0.2.1        pkgconfig_2.0.3        </span></span>
<span><span class="co">##  [94] gtable_0.3.4            blob_1.2.4              impute_1.74.1          </span></span>
<span><span class="co">##  [97] XVector_0.40.0          survMisc_0.5.6          shadowtext_0.1.2       </span></span>
<span><span class="co">## [100] clusterProfiler_4.8.3   htmltools_0.5.7         carData_3.0-5          </span></span>
<span><span class="co">## [103] fgsea_1.26.0            clue_0.3-65             scales_1.2.1           </span></span>
<span><span class="co">## [106] Biobase_2.60.0          png_0.1-8               ggfun_0.1.3            </span></span>
<span><span class="co">## [109] km.ci_0.5-6             rjson_0.2.21            nlme_3.1-163           </span></span>
<span><span class="co">## [112] org.Hs.eg.db_3.17.0     cachem_1.0.8            zoo_1.8-12             </span></span>
<span><span class="co">## [115] GlobalOptions_0.1.2     stringr_1.5.0           parallel_4.3.1         </span></span>
<span><span class="co">## [118] HDO.db_0.99.1           AnnotationDbi_1.62.2    desc_1.4.2             </span></span>
<span><span class="co">## [121] proxyC_0.3.3            pillar_1.9.0            vctrs_0.6.4            </span></span>
<span><span class="co">## [124] slam_0.1-50             car_3.1-2               xtable_1.8-4           </span></span>
<span><span class="co">## [127] cluster_2.1.4           evaluate_0.22           magick_2.8.0           </span></span>
<span><span class="co">## [130] cli_3.6.2               compiler_4.3.1          rlang_1.1.2            </span></span>
<span><span class="co">## [133] crayon_1.5.2            ggsignif_0.6.4          labeling_0.4.3         </span></span>
<span><span class="co">## [136] mclust_6.0.0            skmeans_0.2-16          plyr_1.8.9             </span></span>
<span><span class="co">## [139] fs_1.6.3                stringi_1.7.12          viridisLite_0.4.2      </span></span>
<span><span class="co">## [142] BiocParallel_1.34.2     munsell_0.5.0           Biostrings_2.68.1      </span></span>
<span><span class="co">## [145] lazyeval_0.2.2          GOSemSim_2.26.1         Matrix_1.6-1.1         </span></span>
<span><span class="co">## [148] patchwork_1.1.3         bit64_4.0.5             KEGGREST_1.40.1        </span></span>
<span><span class="co">## [151] gridtext_0.1.5          igraph_1.6.0            broom_1.0.5            </span></span>
<span><span class="co">## [154] memoise_2.0.1           RcppParallel_5.1.7      bslib_0.6.1            </span></span>
<span><span class="co">## [157] ggtree_3.8.2            fastmatch_1.1-4         bit_4.0.5              </span></span>
<span><span class="co">## [160] downloader_0.4          gson_0.1.0              ape_5.7-1</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Zuguang Gu.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  <style>nav[data-toggle='toc'] .nav .nav {display: block;}</style>
</body>
</html>
