<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Automatic Functional Enrichment on Signature Genes • cola</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Automatic Functional Enrichment on Signature Genes">
<meta property="og:description" content="cola">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">cola</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.9.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Standard consensus partitioning</li>
    <li>
      <a href="../articles/cola_quick.html">A Quick Start of Using cola Package</a>
    </li>
    <li>
      <a href="../articles/cola_general.html">A Framework for Consensus Partitioning</a>
    </li>
    <li>
      <a href="../articles/functional_enrichment.html">Automatic Functional Enrichment on Signature genes</a>
    </li>
    <li>
      <a href="../articles/predict.html">Predict Classes for New Samples</a>
    </li>
    <li>
      <a href="../articles/work_with_big_datasets.html">Work with Big Datasets</a>
    </li>
    <li>
      <a href="../articles/compare_partitions.html">Compare Partitions</a>
    </li>
    <li>
      <a href="../articles/ATC_methods.html">ATC - More Forms</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Hierarchical partitioning</li>
    <li>
      <a href="../articles/hierarchical.html">Hierarchical consensus partitioning</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Other examples</li>
    <li>
      <a href="../articles/AML.html">Re-analyze an AML proteomics dataset</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">On public datasets</li>
    <li>
      <a href="../articles/examples.html">Applications on public datasets</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Supplementary files</li>
    <li>
      <a href="https://jokergoo.github.io/cola_supplementary/" class="external-link">Supplementary files for cola NAR paper</a>
    </li>
    <li>
      <a href="https://cola-rh.github.io/supplementary/" class="external-link">Supplementary files for cola BIB paper</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/jokergoo/cola/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Automatic Functional Enrichment on Signature Genes</h1>
                        <h4 data-toc-skip class="author">Zuguang Gu (<a href="mailto:z.gu@dkfz.de" class="email">z.gu@dkfz.de</a>)</h4>
            
            <h4 data-toc-skip class="date">2024-02-27</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/jokergoo/cola/blob/HEAD/vignettes/functional_enrichment.Rmd" class="external-link"><code>vignettes/functional_enrichment.Rmd</code></a></small>
      <div class="hidden name"><code>functional_enrichment.Rmd</code></div>

    </div>

    
    
<p>If the matrix rows can correspond to genes (e.g. the gene expression matrix, or the methylation array data where CpG sites can be annotated to the transcription start site of genes), <em>cola</em> performs functional enrichment by the <code><a href="../reference/functional_enrichment-dispatch.html">functional_enrichment()</a></code> function to the signatures by <a href="https://bioconductor.org/packages/clusterProfiler/" class="external-link"><em>ClusterProfiler</em></a>, <a href="https://bioconductor.org/packages/DOSE/" class="external-link"><em>DOSE</em></a> or <a href="https://bioconductor.org/packages/ReactomePA/" class="external-link"><em>ReactomePA</em></a> packages.</p>
<div class="section level2">
<h2 id="the-first-example">The first example<a class="anchor" aria-label="anchor" href="#the-first-example"></a>
</h2>
<p>We first demonstrate the usage of <code><a href="../reference/functional_enrichment-dispatch.html">functional_enrichment()</a></code> function by the TCGA GBM dataset. In following example code, <code>TCGA_GBM_subgroup.rds</code> is generated by the code demonstrated <a href="https://jokergoo.github.io/cola_examples/TCGA_GBM/" class="external-link">here</a>. We download the result file that has already been generated.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://jokergoo.github.io/cola_examples/TCGA_GBM/TCGA_GBM_subgroup.rds"</span>, </span>
<span>    destfile <span class="op">=</span> <span class="st">"TCGA_GBM_subgroup.rds"</span>, quiet <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">rl</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"TCGA_GBM_subgroup.rds"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.remove</a></span><span class="op">(</span><span class="st">"TCGA_GBM_subgroup.rds"</span><span class="op">)</span></span></code></pre></div>
<p>We select result from a single method <code>ATC:skmeans</code>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/cola" class="external-link">cola</a></span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">=</span> <span class="va">rl</span><span class="op">[</span><span class="st">"ATC:skmeans"</span><span class="op">]</span></span>
<span><span class="va">res</span></span></code></pre></div>
<pre><code><span><span class="co">## A 'ConsensusPartition' object with k = 2, 3, 4, 5, 6.</span></span>
<span><span class="co">##   On a matrix with 11268 rows and 173 columns.</span></span>
<span><span class="co">##   Top rows (1000, 2000, 3000, 4000, 5000) are extracted by 'ATC' method.</span></span>
<span><span class="co">##   Subgroups are detected by 'skmeans' method.</span></span>
<span><span class="co">##   Performed in total 1250 partitions by row resampling.</span></span>
<span><span class="co">##   Best k for subgroups seems to be 4.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Following methods can be applied to this 'ConsensusPartition' object:</span></span>
<span><span class="co">##  [1] "cola_report"             "collect_classes"         "collect_plots"          </span></span>
<span><span class="co">##  [4] "collect_stats"           "colnames"                "compare_partitions"     </span></span>
<span><span class="co">##  [7] "compare_signatures"      "consensus_heatmap"       "dimension_reduction"    </span></span>
<span><span class="co">## [10] "functional_enrichment"   "get_anno"                "get_anno_col"           </span></span>
<span><span class="co">## [13] "get_classes"             "get_consensus"           "get_matrix"             </span></span>
<span><span class="co">## [16] "get_membership"          "get_param"               "get_signatures"         </span></span>
<span><span class="co">## [19] "get_stats"               "is_best_k"               "is_stable_k"            </span></span>
<span><span class="co">## [22] "membership_heatmap"      "ncol"                    "nrow"                   </span></span>
<span><span class="co">## [25] "plot_ecdf"               "predict_classes"         "rownames"               </span></span>
<span><span class="co">## [28] "select_partition_number" "show"                    "suggest_best_k"         </span></span>
<span><span class="co">## [31] "test_to_known_factors"   "top_rows_heatmap"</span></span></code></pre>
<p>We check how the signature genes looks like under 4-group classification:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/get_signatures-dispatch.html">get_signatures</a></span><span class="op">(</span><span class="va">res</span>, k <span class="op">=</span> <span class="fl">4</span>, row_km <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## * 167/173 samples (in 4 classes) remain after filtering by silhouette (&gt;= 0.5).</span></span>
<span><span class="co">## * cache hash: ec46c6ee467d73c0d2f61e5cc6e87c71 (seed 888).</span></span>
<span><span class="co">## * calculating row difference between subgroups by Ftest.</span></span>
<span><span class="co">## * split rows into 4 groups by k-means clustering.</span></span>
<span><span class="co">## * 10023 signatures (89.0%) under fdr &lt; 0.05, group_diff &gt; 0.</span></span>
<span><span class="co">##   - randomly sample 2000 signatures.</span></span>
<span><span class="co">## * making heatmaps for signatures.</span></span></code></pre>
<p><img src="functional_enrichment_files/figure-html/unnamed-chunk-5-1.png" width="500" style="display: block; margin: auto;"></p>
<p>Rows are split into four groups with different expression patterns among samples. The functional enrichment will be applied to genes in each row-cluster.</p>
<p>To apply functional enrichment, the important thing is to check the gene ID type in the input matrix. The helper function <code><a href="../reference/rownames-dispatch.html">rownames()</a></code> directly returns the row names of the matrix stored in <code>res</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/rownames-dispatch.html">rownames</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "AACS"    "FSTL1"   "ELMO2"   "CREB3L1" "RPS11"   "PNMA1"</span></span></code></pre>
<p>The gene ID is symbol. For all enrichment analysis provided by <em>ClusterProfiler</em>, <em>DOSE</em> or <em>ReactomePA</em>, the core ID type is Entrez ID, thus we need to convert from symbol to Entrez ID.</p>
<p>To make it easy, <em>cola</em> automatically tests the gene IDs types and it automatically recognizes three ID types of Ensembl ID, RefSeq ID and gene symbol, which covers most cases of the analysis. If user’s gene ID type is one of the three supported ones, simply run <code><a href="../reference/functional_enrichment-dispatch.html">functional_enrichment()</a></code> on <code>res</code> only with specifying the number of subgroups.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lt</span> <span class="op">=</span> <span class="fu"><a href="../reference/functional_enrichment-dispatch.html">functional_enrichment</a></span><span class="op">(</span><span class="va">res</span>, k <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<p>By default, <code><a href="../reference/functional_enrichment-dispatch.html">functional_enrichment()</a></code> runs enrichment on Gene Ontology, biological function ontologies. <code>ontology</code> can be set as follows:</p>
<ul>
<li>
<code>BP</code>/<code>MF</code>/<code>CC</code>, <code>org_db</code> argument should be set to <a href="http://bioconductor.org/packages/release/BiocViews.html#___Organism" class="external-link">the corresponding database</a>, such as <code>"org.Hs.eg.db"</code>,</li>
<li>
<code>KEGG</code>, <code>organism</code> argument should be set to corresponding species abbreviation, such as <code>"hsa"</code>,</li>
<li>
<code>DO</code>, only works for human,</li>
<li>
<code>MSigDb</code>, only works for human, the path of gmt file should be specified by <code>gmt_file</code> argument. You should only use the gmt files where genes are annotated with the Entrez IDs.</li>
<li>
<code>Reactome</code>, <code>organism</code> argument should be set to the corresponding species, such as <code>"human"</code>.</li>
</ul>
<p><code>ontology</code> can be set as a vector of multiple ontologies.</p>
<p>The value of <code>lt</code> is a list of data frames for different ontologies combined with different k-means groups. Since k-means clustering has already been applied in previous <code><a href="../reference/get_signatures-dispatch.html">get_signatures()</a></code>, the k-means clustering result is stored in <code>res</code> object and <code><a href="../reference/functional_enrichment-dispatch.html">functional_enrichment()</a></code> directly uses the grouping from it.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">lt</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "BP_km1" "BP_km2" "BP_km3" "BP_km4"</span></span></code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">lt</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                    ID                               Description GeneRatio   BgRatio       pvalue</span></span>
<span><span class="co">## GO:0010469 GO:0010469 regulation of signaling receptor activity  116/1443 588/17913 5.094534e-20</span></span>
<span><span class="co">## GO:0008015 GO:0008015                         blood circulation  100/1443 523/17913 1.953545e-16</span></span>
<span><span class="co">## GO:0003013 GO:0003013                circulatory system process  101/1443 532/17913 2.256348e-16</span></span>
<span><span class="co">## GO:0010817 GO:0010817              regulation of hormone levels   95/1443 503/17913 2.560685e-15</span></span>
<span><span class="co">## GO:0030855 GO:0030855           epithelial cell differentiation  123/1443 746/17913 8.436502e-15</span></span>
<span><span class="co">## GO:0055080 GO:0055080                        cation homeostasis  102/1443 588/17913 7.326199e-14</span></span>
<span><span class="co">##                p.adjust       qvalue</span></span>
<span><span class="co">## GO:0010469 2.888091e-16 2.279670e-16</span></span>
<span><span class="co">## GO:0008015 4.263745e-13 3.365521e-13</span></span>
<span><span class="co">## GO:0003013 4.263745e-13 3.365521e-13</span></span>
<span><span class="co">## GO:0010817 3.629131e-12 2.864598e-12</span></span>
<span><span class="co">## GO:0030855 9.565306e-12 7.550226e-12</span></span>
<span><span class="co">## GO:0055080 6.922037e-11 5.463802e-11</span></span>
<span><span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      geneID</span></span>
<span><span class="co">## GO:0010469                                                             BMP3/CCL19/BUD31/GDF10/CGA/NENF/IL23A/WNT10B/CCL23/SCT/VIP/CALCR/EPO/TNFSF15/CCL17/SHANK1/CRLF1/TNFSF18/MUC4/LY6H/TNFSF9/INSL3/OXT/WNT8B/FGF17/RASGRF1/KL/TNF/CAV3/RAMP3/NTS/XCL1/WNT1/CALCB/AMBN/IFNW1/NTF3/FGF5/INHA/FGF21/IL21/DKKL1/CBLC/IFNA21/AVP/CTF1/ADH7/EDN3/GH2/CSF3/ADRB2/IL3/CCL24/ADRA2B/GRIN1/CACNG3/CSH2/LTA/NRG1/IFNA8/GHRH/GDF3/LEP/CCL22/CRHBP/SHH/THPO/GRIN2A/FGF22/CCL11/IAPP/GRP/IL17B/POMC/IL26/PRLH/CDC42EP2/GREM1/TTR/CCL16/CARTPT/TDGF1/IL11/WNT4/TSHB/TNFSF14/PF4/MLN/BMP6/GDF2/F2/IL13/NRTN/IL19/PPY/TNFSF11/OSGIN1/IL4/CCL1/PHLDA2/REG1A/AGTR2/ADRA2C/FGF20/PSPN/CCL27/BMP4/RLN2/BTC/ADM2/IL2/WNT7A/AGRP/RGS9/EDN2/CSF2</span></span>
<span><span class="co">## GO:0008015                                                                                                                         CD38/NPR1/CACNB1/DBH/ADRA1A/FXYD2/CMA1/TACR2/SNTA1/CACNA1C/P2RX1/HBB/CASR/HCN4/ATP2B3/EPO/TH/MYL1/S100A1/CACNA2D3/LPA/FGA/CACNA1S/CASQ2/OXT/CSRP3/CAV3/NTS/CALM2/CASQ1/KCNJ12/KCNJ3/FXYD1/HSPB7/AVP/DRD1/MYL4/NFE2/SLC6A4/KEL/ADORA2A/EDN3/KCNG2/SERPINF2/AZU1/RNF207/ADRA2B/CACNG3/KCNK15/RYR2/GSTO1/NTSR1/LEP/MYH6/TACR1/TNNT2/AGER/CYP4F12/PKP2/SOD1/BBS4/P2RX2/ID2/POMC/KCNE1/DRD5/FGB/COL4A3/GLP1R/P2RY2/CTSG/KCNJ4/CARTPT/NOX1/MYL2/IRX5/GPR37L1/RYR1/ADRB1/ATP2A3/FOXC2/MYL3/CYP4F2/TRPV4/SGCD/REN/AGTR2/ADRA2C/SRI/TNNI3/TNNC1/ADRA1B/DES/ATP1A2/ADM2/HTR1B/IL2/EDN2/FXYD3/GRIP2</span></span>
<span><span class="co">## GO:0003013                                                                                                                   CD38/NPR1/CACNB1/DBH/ADRA1A/FXYD2/CMA1/TACR2/SNTA1/CACNA1C/P2RX1/HBB/CASR/HCN4/ATP2B3/EPO/TH/MYL1/S100A1/CACNA2D3/LPA/FGA/CACNA1S/CASQ2/OXT/CSRP3/CAV3/RAMP3/NTS/CALM2/CASQ1/KCNJ12/KCNJ3/FXYD1/HSPB7/AVP/DRD1/MYL4/NFE2/SLC6A4/KEL/ADORA2A/EDN3/KCNG2/SERPINF2/AZU1/RNF207/ADRA2B/CACNG3/KCNK15/RYR2/GSTO1/NTSR1/LEP/MYH6/TACR1/TNNT2/AGER/CYP4F12/PKP2/SOD1/BBS4/P2RX2/ID2/POMC/KCNE1/DRD5/FGB/COL4A3/GLP1R/P2RY2/CTSG/KCNJ4/CARTPT/NOX1/MYL2/IRX5/GPR37L1/RYR1/ADRB1/ATP2A3/FOXC2/MYL3/CYP4F2/TRPV4/SGCD/REN/AGTR2/ADRA2C/SRI/TNNI3/TNNC1/ADRA1B/DES/ATP1A2/ADM2/HTR1B/IL2/EDN2/FXYD3/GRIP2</span></span>
<span><span class="co">## GO:0010817                                                                                                                                     CD38/TFR2/CYP1A1/CGA/TFAP2B/GFI1/CMA1/TACR2/BAIAP3/CACNA1C/SLC5A7/VIP/CASR/KCNC2/SLC44A4/VSNL1/ALDH1A2/SULT1A1/RDH16/ACHE/DHRS9/HNF1A/FGA/PAX8/CYP26A1/PON3/ARL2/CPLX3/OPRK1/SRD5A2/RDH8/TNF/STAR/GHRHR/INHA/DKKL1/HSD17B6/DIO3/HPN/ADORA2A/EDN3/KCNG2/DUOX1/SULT1E1/SERPINA7/ADRA2B/RAPGEF3/NR1H4/FDXR/HSD17B8/GHRH/CYP11A1/LEP/ISL1/CRHBP/TACR1/SHH/BLK/DIO1/GRP/KLK6/HTR2C/FGB/GLP1R/TTR/CTSG/CARTPT/CFTR/SIRT3/IL11/WNT4/AKR1C3/APOA1/TSHB/CYP3A4/HSD17B2/FOXA2/GCK/BMP6/AKR1C4/CRABP2/GATA3/TNFSF11/PDX1/TRPV4/NR0B2/REN/AGTR2/ADRA2C/CRYM/SRI/DUOX2/RBP4/CYP46A1/DHRS2</span></span>
<span><span class="co">## GO:0030855 XDH/CYP1A1/KRT15/HOXB5/UPK1B/DSC1/KRT34/WNT10B/KRT4/PAX6/CLDN3/BARX1/PGR/SH3BP1/KRTAP2-4/PKP3/SLC44A4/KRT13/KRTAP5-9/KRT24/TGM3/SFN/OVOL2/DHRS9/CLDN5/PAX8/GSTM3/RARG/FOXN1/ELF5/PITX3/KRT5/KRT37/PPL/TNF/KRT8/GLI1/TCF15/PDZD7/WNT1/AQP3/KRTAP5-8/WNT11/TST/ELF3/CDX2/MSX2/TBX1/RHCG/EMX1/LHX1/CCNO/UPK3A/RAPGEF3/NRG1/KRT32/GDF3/ETV2/TCF21/SIX2/PKP2/WNT16/KRT86/PROC/SOD1/KRT16/ID2/KRT35/KRT33A/S100A7/SOX17/GREM1/MGMT/ACTL8/KRT14/PRSS8/WNT4/AKR1C3/KRT23/B9D1/KRT1/FOXJ1/RHEB/KRT17/BMP6/GDF2/BFSP2/KRT83/KLK12/RAB25/PAX2/IL13/CDH3/VIL1/GATA3/FOXC2/PDX1/KRT7/KRT19/KRT12/SULT2B1/IVL/EVPL/ROS1/EHF/PKP1/KRT3/HOXB13/SIX3/TGM5/DLX6/WT1/BMP4/SCEL/CRYGB/KRT81/KLK13/DMBT1/WNT7A/SPDEF/SPRR3/TP63/PTCH2</span></span>
<span><span class="co">## GO:0055080                                                                                                    CCL19/CD38/EPB42/NPR1/TFR2/TFAP2B/ADRA1A/GPR157/SNCA/CACNA1C/CCL23/TMPRSS3/CALCR/P2RX1/CASR/ATP2B3/EPO/GRIN2C/GPR12/RHAG/SLC26A4/CASQ2/VPS33A/OXT/KL/CSRP3/CAV3/TRPA1/GRM1/RAMP3/CALM2/XCL1/CASQ1/CALCB/AVP/DRD1/KEL/ADORA2A/GCM2/SLC9A7/LCK/EDN3/S100A14/CCKBR/RHCG/GRIN1/UPK3A/CD19/RYR2/ATP12A/GSTO1/SCNN1A/TTPA/NTSR1/XK/CLDN16/TACR1/JPH2/FIS1/CYP4F12/SCNN1B/GRIN2A/ATP6V1F/CCL11/IAPP/SOD1/GPR6/P2RX2/GATA2/ALAS2/CALB1/HTR2C/DRD5/TMPRSS6/GLP1R/MS4A1/CFTR/NOX1/LCN2/FZD9/RYR1/BMP6/GDF2/F2/IL13/TRPC5/PLA2G1B/TNFSF11/CYP4F2/TRPV4/CCL1/REG1A/SGCD/AGTR2/SRI/TNNI3/P2RX5/ATP1A2/HTR1B/IL2/NEDD8/EDN2</span></span>
<span><span class="co">##            Count</span></span>
<span><span class="co">## GO:0010469   116</span></span>
<span><span class="co">## GO:0008015   100</span></span>
<span><span class="co">## GO:0003013   101</span></span>
<span><span class="co">## GO:0010817    95</span></span>
<span><span class="co">## GO:0030855   123</span></span>
<span><span class="co">## GO:0055080   102</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="manually-providing-gene-id-mappings">Manually providing gene ID mappings<a class="anchor" aria-label="anchor" href="#manually-providing-gene-id-mappings"></a>
</h2>
<p>If the gene ID type is not any of Ensembl ID, RefSeq ID or gene symbol, user needs to provide a named vector which provides mapping between user’s ID types to Entrez IDs.</p>
<p>In following example we demonstrate how to properly set the ID mapping by <a href="https://jokergoo.github.io/cola_examples/Golub_leukemia/" class="external-link">the Golub leukemia dataset</a>. The result file is already generated and integrate in <em>cola</em> package.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">golub_cola</span><span class="op">)</span> </span></code></pre></div>
<p>To simplify, we only take result from one method:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">=</span> <span class="va">golub_cola</span><span class="op">[</span><span class="st">"ATC:skmeans"</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/rownames-dispatch.html">rownames</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "AFFX-BioC-5_at"  "AFFX-CreX-3_at"  "AFFX-BioDn-3_st" "hum_alu_at"      "AFFX-DapX-M_at" </span></span>
<span><span class="co">## [6] "AFFX-LysX-3_at"</span></span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/get_signatures-dispatch.html">get_signatures</a></span><span class="op">(</span><span class="va">res</span>, k <span class="op">=</span> <span class="fl">3</span>, row_km <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## * 72/72 samples (in 3 classes) remain after filtering by silhouette (&gt;= 0.5).</span></span>
<span><span class="co">## * cache hash: 87c63a8cabc898f97a024514962787f7 (seed 888).</span></span>
<span><span class="co">## * calculating row difference between subgroups by Ftest.</span></span>
<span><span class="co">## * split rows into 4 groups by k-means clustering.</span></span>
<span><span class="co">## * 2058 signatures (50.0%) under fdr &lt; 0.05, group_diff &gt; 0.</span></span>
<span><span class="co">##   - randomly sample 2000 signatures.</span></span>
<span><span class="co">## * making heatmaps for signatures.</span></span></code></pre>
<p><img src="functional_enrichment_files/figure-html/unnamed-chunk-11-1.png" width="500" style="display: block; margin: auto;"></p>
<p>The Golub leukemia dataset is a microarray dataset where the gene ID is the probe ID. Thankfully, there is already an annotation package from Bioconductor (<a href="https://bioconductor.org/packages/hu6800.db/" class="external-link">hu6800.db</a>) that provides mapping between the probe ID to Entrez ID.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">hu6800.db</span><span class="op">)</span></span>
<span><span class="va">x</span> <span class="op">=</span> <span class="va">hu6800ENTREZID</span></span>
<span><span class="va">mapped_probes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/Bimap-keys.html" class="external-link">mappedkeys</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span><span class="va">id_mapping</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">as.list</a></span><span class="op">(</span><span class="va">x</span><span class="op">[</span><span class="va">mapped_probes</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">id_mapping</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     A28102_at   AB000114_at   AB000115_at   AB000220_at AB000381_s_at   AB000409_at </span></span>
<span><span class="co">##        "2556"        "4958"       "10964"       "10512"        "2765"        "8569"</span></span></code></pre>
<p>Proportion of probe IDs that can be mapped:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">id_mapping</span><span class="op">[</span><span class="fu"><a href="../reference/rownames-dispatch.html">rownames</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="../reference/nrow-dispatch.html">nrow</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.8790087</span></span></code></pre>
<p>As you see, the format of <code>id_mapping</code> is simple. Names of the vector are the probe IDs and the values are the Entrez IDs. We can directly assign the ID mapping variable to <code>id_mapping</code> argument.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lt</span> <span class="op">=</span> <span class="fu"><a href="../reference/functional_enrichment-dispatch.html">functional_enrichment</a></span><span class="op">(</span><span class="va">res</span>, k <span class="op">=</span> <span class="fl">3</span>, id_mapping <span class="op">=</span> <span class="va">id_mapping</span><span class="op">)</span></span></code></pre></div>
<p><code><a href="../reference/functional_enrichment-dispatch.html">functional_enrichment()</a></code> can also be applied to two other classes of objects:</p>
<ul>
<li>The <code>ConsensusPartitionList</code> object which is generated by <code>run_all_partition_methods()</code> function. The result is a list (for each method) of lists (for each ontology) of data frames.</li>
<li>Simply a vector of gene IDs. The result is a data frame.</li>
</ul>
</div>
<div class="section level2">
<h2 id="visualize-functional-enrichment-results">Visualize functional enrichment results<a class="anchor" aria-label="anchor" href="#visualize-functional-enrichment-results"></a>
</h2>
<p>The functional enrichment results can be visualized with the <strong>simplifyEnrichment</strong> package. An example plot is as follows:</p>
<p><img src="https://jokergoo.github.io/simplifyEnrichment/articles/simplifyEnrichment_files/figure-html/unnamed-chunk-18-1.png" width="100%"></p>
<p>It clusters function terms based on their pairwise similarities and word cloud annotation is attached to the heatmap to summarize the basic biological functions in each cluster. Users can find more details of the <strong>simplifyEnrichment</strong> analysis from <a href="https://jokergoo.github.io/simplifyEnrichment/articles/simplifyEnrichment.html#apply-to-multiple-lists-of-go-ids-1" class="external-link uri">https://jokergoo.github.io/simplifyEnrichment/articles/simplifyEnrichment.html#apply-to-multiple-lists-of-go-ids-1</a>.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">## Platform: x86_64-apple-darwin20 (64-bit)</span></span>
<span><span class="co">## Running under: macOS Ventura 13.2.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Random number generation:</span></span>
<span><span class="co">##  RNG:     L'Ecuyer-CMRG </span></span>
<span><span class="co">##  Normal:  Inversion </span></span>
<span><span class="co">##  Sample:  Rejection </span></span>
<span><span class="co">##  </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] C/UTF-8/C/C/C/C</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Europe/Berlin</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] hu6800.db_3.13.0     org.Hs.eg.db_3.17.0  AnnotationDbi_1.62.2 IRanges_2.36.0      </span></span>
<span><span class="co">##  [5] S4Vectors_0.40.2     Biobase_2.60.0       BiocGenerics_0.48.1  cola_2.9.1          </span></span>
<span><span class="co">##  [9] knitr_1.44           markdown_1.10       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] blob_1.2.4              bitops_1.0-7            Biostrings_2.68.1       RCurl_1.98-1.12        </span></span>
<span><span class="co">##  [5] fastmap_1.1.1           XML_3.99-0.14           digest_0.6.33           lifecycle_1.0.4        </span></span>
<span><span class="co">##  [9] cluster_2.1.4           Cairo_1.6-2             survival_3.5-8          KEGGREST_1.40.1        </span></span>
<span><span class="co">## [13] RSQLite_2.3.1           magrittr_2.0.3          genefilter_1.82.1       compiler_4.3.1         </span></span>
<span><span class="co">## [17] rlang_1.1.2             sass_0.4.8              rngtools_1.5.2          tools_4.3.1            </span></span>
<span><span class="co">## [21] yaml_2.3.7              brew_1.0-8              doRNG_1.8.6             bit_4.0.5              </span></span>
<span><span class="co">## [25] mclust_6.0.0            xml2_1.3.6              eulerr_7.0.0            RColorBrewer_1.1-3     </span></span>
<span><span class="co">## [29] purrr_1.0.2             desc_1.4.2              grid_4.3.1              xtable_1.8-4           </span></span>
<span><span class="co">## [33] colorspace_2.1-0        iterators_1.0.14        cli_3.6.2               rmarkdown_2.25         </span></span>
<span><span class="co">## [37] crayon_1.5.2            ragg_1.2.6              httr_1.4.7              rjson_0.2.21           </span></span>
<span><span class="co">## [41] DBI_1.1.3               cachem_1.0.8            stringr_1.5.0           splines_4.3.1          </span></span>
<span><span class="co">## [45] zlibbioc_1.46.0         parallel_4.3.1          impute_1.74.1           XVector_0.40.0         </span></span>
<span><span class="co">## [49] matrixStats_1.2.0       vctrs_0.6.4             Matrix_1.6-1.1          jsonlite_1.8.8         </span></span>
<span><span class="co">## [53] slam_0.1-50             GetoptLong_1.0.5        bit64_4.0.5             irlba_2.3.5.1          </span></span>
<span><span class="co">## [57] clue_0.3-65             magick_2.8.0            systemfonts_1.0.5       foreach_1.5.2          </span></span>
<span><span class="co">## [61] jquerylib_0.1.4         annotate_1.78.0         glue_1.6.2              pkgdown_2.0.7          </span></span>
<span><span class="co">## [65] codetools_0.2-19        stringi_1.7.12          shape_1.4.6             GenomeInfoDb_1.36.4    </span></span>
<span><span class="co">## [69] ComplexHeatmap_2.18.0   htmltools_0.5.7         GenomeInfoDbData_1.2.10 circlize_0.4.15        </span></span>
<span><span class="co">## [73] R6_2.5.1                microbenchmark_1.4.10   textshaping_0.3.7       doParallel_1.0.17      </span></span>
<span><span class="co">## [77] rprojroot_2.0.3         evaluate_0.22           lattice_0.21-9          png_0.1-8              </span></span>
<span><span class="co">## [81] memoise_2.0.1           bslib_0.6.1             Rcpp_1.0.11             xfun_0.40              </span></span>
<span><span class="co">## [85] pkgconfig_2.0.3         fs_1.6.3                MatrixGenerics_1.12.3   skmeans_0.2-16         </span></span>
<span><span class="co">## [89] GlobalOptions_0.1.2</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Zuguang Gu.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  <style>nav[data-toggle='toc'] .nav .nav {display: block;}</style>
</body>
</html>
