<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Hierarchical Consensus Partitioning • cola</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Hierarchical Consensus Partitioning">
<meta property="og:description" content="cola">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">cola</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.9.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Standard consensus partitioning</li>
    <li>
      <a href="../articles/cola_quick.html">A Quick Start of Using cola Package</a>
    </li>
    <li>
      <a href="../articles/cola_general.html">A Framework for Consensus Partitioning</a>
    </li>
    <li>
      <a href="../articles/functional_enrichment.html">Automatic Functional Enrichment on Signature genes</a>
    </li>
    <li>
      <a href="../articles/predict.html">Predict Classes for New Samples</a>
    </li>
    <li>
      <a href="../articles/work_with_big_datasets.html">Work with Big Datasets</a>
    </li>
    <li>
      <a href="../articles/compare_partitions.html">Compare Partitions</a>
    </li>
    <li>
      <a href="../articles/ATC_methods.html">ATC - More Forms</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Hierarchical partitioning</li>
    <li>
      <a href="../articles/hierarchical.html">Hierarchical consensus partitioning</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Other examples</li>
    <li>
      <a href="../articles/AML.html">Re-analyze an AML proteomics dataset</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">On public datasets</li>
    <li>
      <a href="../articles/examples.html">Applications on public datasets</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Supplementary files</li>
    <li>
      <a href="https://jokergoo.github.io/cola_supplementary/" class="external-link">Supplementary files for cola NAR paper</a>
    </li>
    <li>
      <a href="https://cola-rh.github.io/supplementary/" class="external-link">Supplementary files for cola BIB paper</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/jokergoo/cola/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Hierarchical Consensus Partitioning</h1>
                        <h4 data-toc-skip class="author">Zuguang Gu (<a href="mailto:z.gu@dkfz.de" class="email">z.gu@dkfz.de</a>)</h4>
            
            <h4 data-toc-skip class="date">2024-02-27</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/jokergoo/cola/blob/HEAD/vignettes/hierarchical.Rmd" class="external-link"><code>vignettes/hierarchical.Rmd</code></a></small>
      <div class="hidden name"><code>hierarchical.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="the-problem">The problem<a class="anchor" aria-label="anchor" href="#the-problem"></a>
</h2>
<p>Choosing the best k (number of partitions) is not an easy problem for consensus partitioning. In consensus partitioning, various metrics based on the consensus matrix are normally calculated, <em>e.g.</em> PAC scores (or 1-PAC) or mean silhouette scores, and the best k is normally selected based on the “extremal method”, <em>i.e.</em> to select the k that corresponds to the highest or lowest values of the metrics. When the number of partitions is small, it is relatively easy to determine the best k with high confidence, while when the real number of clusters gets very large, it is difficult to identify the correct or approximate k for several reasons, some of which we list in the following: 1) Variation in the “big clusters” affect the eCDF of the consensus matrix stronger than variation in the “small clusters”, this can strongly affect PAC scores. 2) Groups showing weaker differences (we can call them secondary groups) are more difficult to separate especially when there are already other groups showing distinct differences (we can call them major groups). 3) The curve of various metrics against k gets flattened for large k and the value of k with the extremal values will be less distinct.</p>
<p>The following four figures illustrate the eCDF curves of a consensus matrix, 1-PAC, mean silhouette and concordance scores for different k where k ranges from 2 to 10 (from the analysis <a href="https://jokergoo.github.io/cola_examples/GBM_450K/GBM_450K_cgi_all_subgroup_cola_report/cola_report.html#MAD-pam" class="external-link">here</a>). Basically, when k &gt;= 5, the eCDF curves have a long plateau with less and less curvature, <em>i.e.</em> they lose their step-like shape, which results in 1-PAC getting almost stable for k &gt;= 5. Also for the curves of mean silhouette and concordance scores against k, they are almost flattened for k &gt;= 3. If using the “extremal method”, k = 6 is taken as the best k because 1-PAC selects k = 9 while mean silhouette and concordance select k = 6.</p>
<p><img src="large_k_problem_select_partition_number-min.png" width="100%"></p>
<p>When inspecting the consensus heatmaps for different k (<em>cf.</em> heatmaps below), actually it is difficult to assess whether the partitioning for k = 6 is better than any other k in [5, 10].</p>
<p><img src="large_k_problem_consensus_heatmap-min.png" width="100%"></p>
<p>The problems of the “big clusters / small clusters” or “major clusters / secondary clusters” in selecting the best k are mainly due to the consensus partitioning procedure that all samples are taken into account equally. From version 2.0.0, we proposed a new method which tries to solve this issue by applying consensus partitioning in a hierarchical way. Simply speaking, one could first classify the samples into n<sub>major</sub> groups (n<sub>major</sub> is a small number, major clusters), then for each subgroup of samples, one could repeatedly apply consensus clustering. By this means, theoretically, small clusters or secondary clusters could be detected in later steps of the hierarchical procedure.</p>
</div>
<div class="section level2">
<h2 id="the-workflow">The workflow<a class="anchor" aria-label="anchor" href="#the-workflow"></a>
</h2>
<p>The figure below illustrates the workflow of the hierarchical consensus partitioning.</p>
<p><img src="hierarchical_partitioning_workflow.png" width="400"></p>
<p>The steps are:</p>
<ol style="list-style-type: decimal">
<li>The input matrix <em>M</em> is treated as the top node in the hierarchy, with a node label “0”.</li>
<li>Apply multiple combinations of top-value methods and partitioning methods on the matrix with a small set of k (<em>e.g.</em>, 2-4).</li>
<li>Select the consensus partitioning result (<em>i.e.</em> with a specific top-value method and a partitioning method) which shows the highest mean silhouette scores with its best k. This consensus partitioning is treated as the partitioning on the node.</li>
<li>If this partitioning is not stable (<em>e.g.</em>, mean silhouette &lt; 0.9), the hierarchical partitioning stops and the columns of the matrix are treated as a subgroup.</li>
<li>If the number of signatures or the fraction of signatures to the total number of rows in the matrix is too small, it means the partitioning does not show biological meaningful results and the hierarchical partitioning stops. The columns are treated as a subgroup.</li>
<li>If the partioning passes the filtering on step 4 and 5, each subgroup is treated a child node of the current node.</li>
<li>For each submatrix, if the number of columns is too small, the hierarchical partitioning stops and the set of columns is treated as a subgroup.</li>
<li>If the submatrix has enough columns, go to step 2 to perform consensus partioning recursively.</li>
</ol>
<p>The process of the hierarchical consensus partitioning is saved as a dendrogram internally.</p>
</div>
<div class="section level2">
<h2 id="the-usage">The usage<a class="anchor" aria-label="anchor" href="#the-usage"></a>
</h2>
<p>In this section, we demonstrate the functionalities of hierarchical consensus partitioning. The design of these new functionalities tries to be as consistent as the functions for normal consensus partitioning in <strong>cola</strong>, <em>i.e.</em>, the function <code><a href="../reference/consensus_partition.html">consensus_partition()</a></code> or <code><a href="../reference/run_all_consensus_partition_methods.html">run_all_consensus_partition_methods()</a></code>. Thus, you may find many functions having the same names as the functions for normal consensus partitioning.</p>
<p>The following code applies hierarchical consensus partitioning on the Golub dataset. Function <code><a href="../reference/hierarchical_partition.html">hierarchical_partition()</a></code> applies the analysis where the main arguments are very similar as in <code><a href="../reference/consensus_partition.html">consensus_partition()</a></code> or <code><a href="../reference/run_all_consensus_partition_methods.html">run_all_consensus_partition_methods()</a></code>, which are the input matrix, the sample annotations and the number of cores. The function returns a <code>HierarchicalPartition</code> object.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">golubEsets</span><span class="op">)</span>  <span class="co"># from Bioconductor</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">Golub_Merge</span><span class="op">)</span></span>
<span><span class="va">m</span> <span class="op">=</span> <span class="fu">exprs</span><span class="op">(</span><span class="va">Golub_Merge</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/colnames-dispatch.html">colnames</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"sample_"</span>, <span class="fu"><a href="../reference/colnames-dispatch.html">colnames</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">anno</span> <span class="op">=</span> <span class="fu">pData</span><span class="op">(</span><span class="va">Golub_Merge</span><span class="op">)</span></span>
<span></span>
<span><span class="va">m</span><span class="op">[</span><span class="va">m</span> <span class="op">&lt;=</span> <span class="fl">1</span><span class="op">]</span> <span class="op">=</span> <span class="cn">NA</span></span>
<span><span class="va">m</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span></span>
<span></span>
<span><span class="va">m</span> <span class="op">=</span> <span class="fu"><a href="../reference/adjust_matrix.html">adjust_matrix</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># apply quantile normalization</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/bmbolstad/preprocessCore" class="external-link">preprocessCore</a></span><span class="op">)</span>  <span class="co"># from Bioconductor</span></span>
<span><span class="va">cn</span> <span class="op">=</span> <span class="fu"><a href="../reference/colnames-dispatch.html">colnames</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span></span>
<span><span class="va">rn</span> <span class="op">=</span> <span class="fu"><a href="../reference/rownames-dispatch.html">rownames</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span></span>
<span><span class="va">m</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/preprocessCore/man/normalize.quantiles.html" class="external-link">normalize.quantiles</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/colnames-dispatch.html">colnames</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span> <span class="op">=</span> <span class="va">cn</span></span>
<span><span class="fu"><a href="../reference/rownames-dispatch.html">rownames</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span> <span class="op">=</span> <span class="va">rn</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">golub_cola_rh</span> <span class="op">=</span> <span class="fu"><a href="../reference/hierarchical_partition.html">hierarchical_partition</a></span><span class="op">(</span></span>
<span>    <span class="va">m</span>, anno <span class="op">=</span> <span class="va">anno</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ALL.AML"</span><span class="op">)</span>, drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Some important arguments in <code><a href="../reference/hierarchical_partition.html">hierarchical_partition()</a></code> are listed as follows:</p>
<ul>
<li>
<code>top_n</code>: Number of rows with top values. The value can be a vector of integers. On each node, there is an additional filtering on rows of the submatrices to remove those rows with very small variances, which results in the reducing number of rows in the submatrices. Here <code>top_n</code> can be set as a vector of values less than 1 which are treated as the fraction of rows of every submatrix.</li>
<li>
<code>top_value_method</code>: A single or a vector of top-value methods.</li>
<li>
<code>partition_method</code>: A single or a vector of partitioning methods. All combinations of <code>top_value_method</code> and <code>partition_method</code> are tried.</li>
<li>
<code>combination_method</code>: Instead of specifying <code>top_value_method</code> and <code>partition_method</code>, all methods that are going to be tried can also be specified with <code>combination_method</code>. The value can be a vector in a form of <code>c("SD:hclust", "ATC:skmeans", ...)</code> or a data frame with two columns <code>data.frame(c("SD", "ATC"), c("hclust", "skmeans"))</code>.</li>
<li>
<code>anno</code>: Annotation for the columns. The value can be a vector or a data frame. Note the rows in <code>anno</code> should be corresponded to the matrix columns.</li>
<li>
<code>anno_col</code>: Colors for annotations. The value should be a list of colors where a named vector for discrete color mapping and a color mapping function generated by <code><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html" class="external-link">circlize::colorRamp2()</a></code> for continuous color mapping.</li>
<li>
<code>mean_silhouette_cutoff</code>: Cut off for mean silhouette_cutoff score to decide whether a partition is stable and to be split further more.</li>
<li>
<code>min_samples</code>: Miminal number of samples to perform partitioning.</li>
<li>
<code>subset</code>: Number of subset columns to perform partitioning. If the current number of columns in the submatrix is higher than <code>subset</code>, <code><a href="../reference/consensus_partition_by_down_sampling.html">consensus_partition_by_down_sampling()</a></code> instead of <code><a href="../reference/consensus_partition.html">consensus_partition()</a></code> will be applied. It will be discussed in more details in the Section <a href="hierarchical.html#toc_5"><strong>Work with huge datasets</strong></a>.</li>
<li>
<code>min_n_signatures</code>: On each node that has a partition, <code><a href="../reference/get_signatures-dispatch.html">get_signatures()</a></code> is applied to find the number of signatures under the best k. If the number of signatures is less than <code>min_n_signatures</code>, it means the partitioning might not be significantly different and the hierarchical consensus partitioning stops.</li>
<li>
<code>min_p_signatures</code>: This is the fraction of signatures to the total number of rows of the original matrix. This filtering is a companion of <code>min_n_signatures</code>.</li>
<li>
<code>max_k</code>: Maximal number of groups to try for consensus partitioning. Normally this value should be set to a small value, because more subgroups will be found during the hierarchical consensus partitioning process.</li>
<li>
<code>cores</code>: <code><a href="../reference/hierarchical_partition.html">hierarchical_partition()</a></code> supports parallel computing. This is the number of cores to use.</li>
</ul>
<p>The object <code>golub_cola_rh</code> is already generated and shipped in <strong>cola</strong> package, so we directly load it.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">golub_cola_rh</span><span class="op">)</span></span>
<span><span class="va">golub_cola_rh</span></span></code></pre></div>
<pre><code><span><span class="co">## A 'HierarchicalPartition' object with 'ATC:skmeans' method.</span></span>
<span><span class="co">##   On a matrix with 4116 rows and 72 columns.</span></span>
<span><span class="co">##   Performed in total 1350 partitions.</span></span>
<span><span class="co">##   There are 6 groups under the following parameters:</span></span>
<span><span class="co">##     - min_samples: 6</span></span>
<span><span class="co">##     - mean_silhouette_cutoff: 0.9</span></span>
<span><span class="co">##     - min_n_signatures: 103 (signatures are selected based on:)</span></span>
<span><span class="co">##       - fdr_cutoff: 0.05</span></span>
<span><span class="co">##       - group_diff (scaled values): 0.5</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Hierarchy of the partition:</span></span>
<span><span class="co">##   0, 72 cols, 2068 signatures</span></span>
<span><span class="co">##   |-- 01, 35 cols, 652 signatures</span></span>
<span><span class="co">##   |   |-- 011, 11 cols (b)</span></span>
<span><span class="co">##   |   |-- 012, 13 cols, 5 signatures (c)</span></span>
<span><span class="co">##   |   `-- 013, 11 cols (b)</span></span>
<span><span class="co">##   |-- 02, 24 cols, 138 signatures</span></span>
<span><span class="co">##   |   |-- 021, 13 cols (a)</span></span>
<span><span class="co">##   |   `-- 022, 11 cols (b)</span></span>
<span><span class="co">##   `-- 03, 13 cols, 7 signatures (c)</span></span>
<span><span class="co">## Stop reason:</span></span>
<span><span class="co">##   a) Mean silhouette score was too small</span></span>
<span><span class="co">##   b) Subgroup had too few columns.</span></span>
<span><span class="co">##   c) There were too few signatures.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Following methods can be applied to this 'HierarchicalPartition' object:</span></span>
<span><span class="co">##  [1] "all_leaves"            "all_nodes"             "cola_report"           "collect_classes"      </span></span>
<span><span class="co">##  [5] "colnames"              "compare_signatures"    "dimension_reduction"   "functional_enrichment"</span></span>
<span><span class="co">##  [9] "get_anno"              "get_anno_col"          "get_children_nodes"    "get_classes"          </span></span>
<span><span class="co">## [13] "get_matrix"            "get_signatures"        "is_leaf_node"          "max_depth"            </span></span>
<span><span class="co">## [17] "merge_node"            "ncol"                  "node_info"             "node_level"           </span></span>
<span><span class="co">## [21] "nrow"                  "rownames"              "show"                  "split_node"           </span></span>
<span><span class="co">## [25] "suggest_best_k"        "test_to_known_factors" "top_rows_heatmap"      "top_rows_overlap"     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## You can get result for a single node by e.g. object["01"]</span></span></code></pre>
<p>Directly entering <code>golub_cola_rh</code> prints the hierarchy. As you already see in the previous output, the node in the hierarchy is encoded in a special way. As explained in previous text, on each node, the columns are split into several groups and a digit (the subgroup index in current partitioning) is appended with the current node label to the children node labels. Thus, the length (or <code>nchar</code>) of the label represents the depth of that node in the hierarchy and from the node label, it is also straightforward to infer its parent node. <em>E.g.</em>, a node with label <code>0123</code> has its parent node <code>012</code>.</p>
<p>Also you can find the functions that can be applied to the <code>HierarchicalPartition</code> object in previous output.</p>
<p>The first function you may try is to see how the columns are separated and the hierarchy of the subgroups. This can be done by <code><a href="../reference/collect_classes-dispatch.html">collect_classes()</a></code> function:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/collect_classes-dispatch.html">collect_classes</a></span><span class="op">(</span><span class="va">golub_cola_rh</span><span class="op">)</span></span></code></pre></div>
<p><img src="hierarchical_files/figure-html/unnamed-chunk-4-1.png" width="700" style="display: block; margin: auto;"></p>
<p>There are several metrics saved for each node which can be retrieved by <code><a href="../reference/node_info-HierarchicalPartition-method.html">node_info()</a></code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/node_info-HierarchicalPartition-method.html">node_info</a></span><span class="op">(</span><span class="va">golub_cola_rh</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    id best_method depth best_k n_columns n_signatures p_signatures is_leaf</span></span>
<span><span class="co">## 1   0 ATC:skmeans     1      3        72         2068  0.502429543   FALSE</span></span>
<span><span class="co">## 2  01 ATC:skmeans     2      3        35          652  0.158406220   FALSE</span></span>
<span><span class="co">## 3 011 not applied     3     NA        11           NA           NA    TRUE</span></span>
<span><span class="co">## 4 012 ATC:skmeans     3      3        13            5  0.001214772    TRUE</span></span>
<span><span class="co">## 5 013 not applied     3     NA        11           NA           NA    TRUE</span></span>
<span><span class="co">## 6  02 ATC:skmeans     2      2        24          138  0.033527697   FALSE</span></span>
<span><span class="co">## 7 021 ATC:skmeans     3      2        13           NA           NA    TRUE</span></span>
<span><span class="co">## 8 022 not applied     3     NA        11           NA           NA    TRUE</span></span>
<span><span class="co">## 9  03 ATC:skmeans     2      2        13            7  0.001700680    TRUE</span></span></code></pre>
<p>There are following columns from <code><a href="../reference/node_info-HierarchicalPartition-method.html">node_info()</a></code>:</p>
<ul>
<li>
<code>id</code>: The node id or label.</li>
<li>
<code>best_method</code>: Because on each node, multiple methods set in <code>combination_method</code> are tried and the method that gives the highest 1-PAC under its best k is finally selected. The best method is saved here.</li>
<li>
<code>depth</code>: The depth of the node in the hierarchy.</li>
<li>
<code>best_k</code>: The best k used.</li>
<li>
<code>n_columns</code>: Number of columns of the submatrix.</li>
<li>
<code>n_signatures</code>: Number of signatures.</li>
<li>
<code>p_signatures</code>: Fraction of signatures to the total number of rows of the matrix.</li>
<li>
<code>is_leaf</code>: Whether the node is a leaf node.</li>
</ul>
<p>These values are useful to merge the children nodes.</p>
<p>Most functions for dealing with the <code>HierarchicalPartition</code> object accept a <code>merge_node</code> argument, where you can set different paremeters to select the children node to merge. These parameters should be set by the function <code><a href="../reference/merge_node_param.html">merge_node_param()</a></code> function. And there are the four parameters can be adjusted:</p>
<ul>
<li><code>depth</code></li>
<li><code>min_n_signatures</code></li>
<li><code>min_p_signatures</code></li>
</ul>
<p><code>min_n_signatures</code> measures how much you trust the classification on the bioloigcal point of view. In the following, we demonstrate to manuplate the dendrogram by setting different <code>min_n_signatures</code> values.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/collect_classes-dispatch.html">collect_classes</a></span><span class="op">(</span><span class="va">golub_cola_rh</span>, merge_node <span class="op">=</span> <span class="fu"><a href="../reference/merge_node_param.html">merge_node_param</a></span><span class="op">(</span>min_n_signatures <span class="op">=</span> <span class="fl">174</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/collect_classes-dispatch.html">collect_classes</a></span><span class="op">(</span><span class="va">golub_cola_rh</span>, merge_node <span class="op">=</span> <span class="fu"><a href="../reference/merge_node_param.html">merge_node_param</a></span><span class="op">(</span>min_n_signatures <span class="op">=</span> <span class="fl">263</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/collect_classes-dispatch.html">collect_classes</a></span><span class="op">(</span><span class="va">golub_cola_rh</span>, merge_node <span class="op">=</span> <span class="fu"><a href="../reference/merge_node_param.html">merge_node_param</a></span><span class="op">(</span>min_n_signatures <span class="op">=</span> <span class="fl">2041</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="hierarchical_files/figure-html/unnamed-chunk-7-1.png" width="672" style="display: block; margin: auto;"></p>
<p>We can also compare to the normal consensus partitioning classification:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">golub_cola</span><span class="op">)</span></span>
<span><span class="va">golub_cola_cp</span> <span class="op">=</span> <span class="va">golub_cola</span><span class="op">[</span><span class="st">"ATC:skmeans"</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/collect_classes-dispatch.html">collect_classes</a></span><span class="op">(</span><span class="va">golub_cola_rh</span>, </span>
<span>    anno <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="fu"><a href="../reference/get_anno-dispatch.html">get_anno</a></span><span class="op">(</span><span class="va">golub_cola_rh</span><span class="op">)</span>, </span>
<span>        cola_cp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="../reference/get_classes-dispatch.html">get_classes</a></span><span class="op">(</span><span class="va">golub_cola_cp</span>, k <span class="op">=</span> <span class="fu"><a href="../reference/suggest_best_k-dispatch.html">suggest_best_k</a></span><span class="op">(</span><span class="va">golub_cola_cp</span><span class="op">)</span><span class="op">)</span><span class="op">[</span>, <span class="st">"class"</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    anno_col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="../reference/get_anno_col-dispatch.html">get_anno_col</a></span><span class="op">(</span><span class="va">golub_cola_rh</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="hierarchical_files/figure-html/unnamed-chunk-8-1.png" width="700" style="display: block; margin: auto;"></p>
<p><code><a href="../reference/get_classes-dispatch.html">get_classes()</a></code> returns the subgroups of columns:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/get_classes-dispatch.html">get_classes</a></span><span class="op">(</span><span class="va">golub_cola_rh</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## sample_39 sample_40 sample_42 sample_47 sample_48 sample_49 sample_41 sample_43 sample_44 sample_45 </span></span>
<span><span class="co">##      "03"      "03"     "013"     "012"     "011"      "03"     "011"     "012"     "012"     "012" </span></span>
<span><span class="co">## sample_46 sample_70 sample_71 sample_72 sample_68 sample_69 sample_67 sample_55 sample_56 sample_59 </span></span>
<span><span class="co">##     "012"     "012"     "013"     "013"     "011"     "011"     "013"      "03"      "03"     "012" </span></span>
<span><span class="co">## sample_52 sample_53 sample_51 sample_50 sample_54 sample_57 sample_58 sample_60 sample_61 sample_65 </span></span>
<span><span class="co">##     "022"     "021"     "021"     "021"     "012"     "022"     "022"     "012"     "022"     "022" </span></span>
<span><span class="co">## sample_66 sample_63 sample_64 sample_62  sample_1  sample_2  sample_3  sample_4  sample_5  sample_6 </span></span>
<span><span class="co">##     "013"     "022"     "022"     "022"      "03"     "013"      "03"      "03"     "011"      "03" </span></span>
<span><span class="co">##  sample_7  sample_8  sample_9 sample_10 sample_11 sample_12 sample_13 sample_14 sample_15 sample_16 </span></span>
<span><span class="co">##      "03"      "03"     "013"     "013"     "013"     "021"     "011"     "013"     "011"     "012" </span></span>
<span><span class="co">## sample_17 sample_18 sample_19 sample_20 sample_21 sample_22 sample_23 sample_24 sample_25 sample_26 </span></span>
<span><span class="co">##     "011"     "012"     "012"     "011"     "011"     "021"      "03"     "011"     "021"     "012" </span></span>
<span><span class="co">## sample_27 sample_34 sample_35 sample_36 sample_37 sample_38 sample_28 sample_29 sample_30 sample_31 </span></span>
<span><span class="co">##      "03"     "022"     "022"     "021"     "021"     "021"     "021"     "013"     "021"     "022" </span></span>
<span><span class="co">## sample_32 sample_33 </span></span>
<span><span class="co">##     "021"     "021"</span></span></code></pre>
<p>Also you can control <code>merge_node</code> argument to decide on which level of hierarchy you want.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/get_classes-dispatch.html">get_classes</a></span><span class="op">(</span><span class="va">golub_cola_rh</span>, merge_node <span class="op">=</span> <span class="fu"><a href="../reference/merge_node_param.html">merge_node_param</a></span><span class="op">(</span>min_n_signatures <span class="op">=</span> <span class="fl">263</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## sample_39 sample_40 sample_42 sample_47 sample_48 sample_49 sample_41 sample_43 sample_44 sample_45 </span></span>
<span><span class="co">##      "03"      "03"     "013"     "012"     "011"      "03"     "011"     "012"     "012"     "012" </span></span>
<span><span class="co">## sample_46 sample_70 sample_71 sample_72 sample_68 sample_69 sample_67 sample_55 sample_56 sample_59 </span></span>
<span><span class="co">##     "012"     "012"     "013"     "013"     "011"     "011"     "013"      "03"      "03"     "012" </span></span>
<span><span class="co">## sample_52 sample_53 sample_51 sample_50 sample_54 sample_57 sample_58 sample_60 sample_61 sample_65 </span></span>
<span><span class="co">##      "02"      "02"      "02"      "02"     "012"      "02"      "02"     "012"      "02"      "02" </span></span>
<span><span class="co">## sample_66 sample_63 sample_64 sample_62  sample_1  sample_2  sample_3  sample_4  sample_5  sample_6 </span></span>
<span><span class="co">##     "013"      "02"      "02"      "02"      "03"     "013"      "03"      "03"     "011"      "03" </span></span>
<span><span class="co">##  sample_7  sample_8  sample_9 sample_10 sample_11 sample_12 sample_13 sample_14 sample_15 sample_16 </span></span>
<span><span class="co">##      "03"      "03"     "013"     "013"     "013"      "02"     "011"     "013"     "011"     "012" </span></span>
<span><span class="co">## sample_17 sample_18 sample_19 sample_20 sample_21 sample_22 sample_23 sample_24 sample_25 sample_26 </span></span>
<span><span class="co">##     "011"     "012"     "012"     "011"     "011"      "02"      "03"     "011"      "02"     "012" </span></span>
<span><span class="co">## sample_27 sample_34 sample_35 sample_36 sample_37 sample_38 sample_28 sample_29 sample_30 sample_31 </span></span>
<span><span class="co">##      "03"      "02"      "02"      "02"      "02"      "02"      "02"     "013"      "02"      "02" </span></span>
<span><span class="co">## sample_32 sample_33 </span></span>
<span><span class="co">##      "02"      "02"</span></span></code></pre>
<p><code><a href="../reference/suggest_best_k-dispatch.html">suggest_best_k()</a></code> extracts the the best k as well as related metrics for the best partitions on each node.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/suggest_best_k-dispatch.html">suggest_best_k</a></span><span class="op">(</span><span class="va">golub_cola_rh</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  node best_method is_leaf best_k 1-PAC mean_silhouette concordance n_sample   </span></span>
<span><span class="co">##     0 ATC:skmeans              3 1.000           0.964       0.984       72 **</span></span>
<span><span class="co">##    01 ATC:skmeans              3 1.000           0.951       0.982       35 **</span></span>
<span><span class="co">##   011 not applied    ✓(b)     NA    NA              NA          NA       11   </span></span>
<span><span class="co">##   012 ATC:skmeans    ✓(c)      3 1.000           0.987       0.994       13 **</span></span>
<span><span class="co">##   013 not applied    ✓(b)     NA    NA              NA          NA       11   </span></span>
<span><span class="co">##    02 ATC:skmeans              4 0.913           0.901       0.944       24  *</span></span>
<span><span class="co">##   021 ATC:skmeans    ✓(a)      2 0.846           0.898       0.965       13   </span></span>
<span><span class="co">##   022 not applied    ✓(b)     NA    NA              NA          NA       11   </span></span>
<span><span class="co">##    03 ATC:skmeans    ✓(c)      2 0.705           0.932       0.971       13   </span></span>
<span><span class="co">## --------------------------------------------------------------------------- </span></span>
<span><span class="co">## Stop reason:</span></span>
<span><span class="co">##   a) Mean silhouette score was too small</span></span>
<span><span class="co">##   b) Subgroup had too few columns.</span></span>
<span><span class="co">##   c) There were too few signatures.</span></span></code></pre>
<p>Another important function which gives a direct feeling of how the subgrouping look like is to check the signatures that are significantly different between subgroups. Similarly as normal consensus partitioning, we can use <code><a href="../reference/get_signatures-dispatch.html">get_signatures()</a></code> here. The function basically applies <code>get_signatures,ConsesusPartition-method()</code> on the partition on every node and collect all the signatures as the signatures of the hierarchical consensus partitioning.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/get_signatures-dispatch.html">get_signatures</a></span><span class="op">(</span><span class="va">golub_cola_rh</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="hierarchical_files/figure-html/unnamed-chunk-12-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Other useful functions are <code><a href="../reference/dimension_reduction-dispatch.html">dimension_reduction()</a></code>, <code><a href="../reference/compare_signatures-dispatch.html">compare_signatures()</a></code> and <code><a href="../reference/test_to_known_factors-dispatch.html">test_to_known_factors()</a></code>. The usages are as follows:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/dimension_reduction-dispatch.html">dimension_reduction</a></span><span class="op">(</span><span class="va">golub_cola_rh</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## use UMAP</span></span></code></pre>
<p><img src="hierarchical_files/figure-html/unnamed-chunk-13-1.png" width="700" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/compare_signatures-dispatch.html">compare_signatures</a></span><span class="op">(</span><span class="va">golub_cola_rh</span><span class="op">)</span></span></code></pre></div>
<p><img src="hierarchical_files/figure-html/unnamed-chunk-14-1.png" width="700" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/test_to_known_factors-dispatch.html">test_to_known_factors</a></span><span class="op">(</span><span class="va">golub_cola_rh</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##           ALL.AML</span></span>
<span><span class="co">## class 5.33797e-09</span></span></code></pre>
<p>Note all these functions mentioned above allow the <code>merge_node</code> argument to adjust the hierarchy.</p>
</div>
<div class="section level2">
<h2 id="automated-reporting">Automated reporting<a class="anchor" aria-label="anchor" href="#automated-reporting"></a>
</h2>
<p>One of the key advantage of <strong>cola</strong> package is it automates the complete analysis. There is also a <code><a href="../reference/cola_report-dispatch.html">cola_report()</a></code> function for <code>HierarchicalPartition</code> class and it automates the complete analysis as well. Simply run:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rh</span> <span class="op">=</span> <span class="fu"><a href="../reference/hierarchical_partition.html">hierarchical_partition</a></span><span class="op">(</span><span class="va">...</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/cola_report-dispatch.html">cola_report</a></span><span class="op">(</span><span class="va">rh</span>, output_dir <span class="op">=</span> <span class="va">...</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="work-with-huge-datasets">Work with huge datasets<a class="anchor" aria-label="anchor" href="#work-with-huge-datasets"></a>
</h2>
<p>In the vignette <a href="work_with_big_datasets.html">“Work with Big Datasets”</a>, we introduced a <code>DownSamplingConsensusPartition</code> class and its corresponding method <code><a href="../reference/consensus_partition_by_down_sampling.html">consensus_partition_by_down_sampling()</a></code> which performs consensus partitioning on a randomly sampled subset of columns and predict the subgroup labels for the remaining columns from the … Here <code><a href="../reference/hierarchical_partition.html">hierarchical_partition()</a></code> also supports down sampling which makes it possible to work on extremly large datasets.</p>
<p>The only thing for dealing with huge datasets is to set the <code>subset</code> argument.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/hierarchical_partition.html">hierarchical_partition</a></span><span class="op">(</span><span class="va">...</span>, subset <span class="op">=</span> <span class="va">...</span><span class="op">)</span></span></code></pre></div>
<p>On each node, to consider the euqal sizes of groups, we first perform a fast k-means and random sample columns with different weight according to the size the groups.</p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">## Platform: x86_64-apple-darwin20 (64-bit)</span></span>
<span><span class="co">## Running under: macOS Ventura 13.2.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Random number generation:</span></span>
<span><span class="co">##  RNG:     L'Ecuyer-CMRG </span></span>
<span><span class="co">##  Normal:  Inversion </span></span>
<span><span class="co">##  Sample:  Rejection </span></span>
<span><span class="co">##  </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] C/UTF-8/C/C/C/C</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Europe/Berlin</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] grid      stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] cola_2.9.1            circlize_0.4.15       ComplexHeatmap_2.18.0 knitr_1.44           </span></span>
<span><span class="co">## [5] markdown_1.10        </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] bitops_1.0-7            DBI_1.1.3               gridExtra_2.3          </span></span>
<span><span class="co">##   [4] rlang_1.1.2             magrittr_2.0.3          clue_0.3-65            </span></span>
<span><span class="co">##   [7] GetoptLong_1.0.5        matrixStats_1.2.0       compiler_4.3.1         </span></span>
<span><span class="co">##  [10] RSQLite_2.3.1           skmeans_0.2-16          polylabelr_0.2.0       </span></span>
<span><span class="co">##  [13] png_0.1-8               systemfonts_1.0.5       vctrs_0.6.4            </span></span>
<span><span class="co">##  [16] stringr_1.5.0           pkgconfig_2.0.3         shape_1.4.6            </span></span>
<span><span class="co">##  [19] crayon_1.5.2            fastmap_1.1.1           XVector_0.40.0         </span></span>
<span><span class="co">##  [22] magick_2.8.0            utf8_1.2.3              rmarkdown_2.25         </span></span>
<span><span class="co">##  [25] ragg_1.2.6              purrr_1.0.2             bit_4.0.5              </span></span>
<span><span class="co">##  [28] xfun_0.40               zlibbioc_1.46.0         cachem_1.0.8           </span></span>
<span><span class="co">##  [31] GenomeInfoDb_1.36.4     jsonlite_1.8.8          blob_1.2.4             </span></span>
<span><span class="co">##  [34] irlba_2.3.5.1           parallel_4.3.1          data.tree_1.1.0        </span></span>
<span><span class="co">##  [37] cluster_2.1.4           R6_2.5.1                bslib_0.6.1            </span></span>
<span><span class="co">##  [40] stringi_1.7.12          RColorBrewer_1.1-3      reticulate_1.34.0      </span></span>
<span><span class="co">##  [43] genefilter_1.82.1       jquerylib_0.1.4         Rcpp_1.0.11            </span></span>
<span><span class="co">##  [46] iterators_1.0.14        IRanges_2.36.0          splines_4.3.1          </span></span>
<span><span class="co">##  [49] Matrix_1.6-1.1          tidyselect_1.2.0        yaml_2.3.7             </span></span>
<span><span class="co">##  [52] viridis_0.6.4           doParallel_1.0.17       codetools_0.2-19       </span></span>
<span><span class="co">##  [55] doRNG_1.8.6             lattice_0.21-9          tibble_3.2.1           </span></span>
<span><span class="co">##  [58] Biobase_2.60.0          KEGGREST_1.40.1         askpass_1.2.0          </span></span>
<span><span class="co">##  [61] evaluate_0.22           survival_3.5-8          desc_1.4.2             </span></span>
<span><span class="co">##  [64] polyclip_1.10-6         xml2_1.3.6              mclust_6.0.0           </span></span>
<span><span class="co">##  [67] Biostrings_2.68.1       pillar_1.9.0            MatrixGenerics_1.12.3  </span></span>
<span><span class="co">##  [70] rngtools_1.5.2          foreach_1.5.2           stats4_4.3.1           </span></span>
<span><span class="co">##  [73] generics_0.1.3          RCurl_1.98-1.12         rprojroot_2.0.3        </span></span>
<span><span class="co">##  [76] S4Vectors_0.40.2        ggplot2_3.4.4           munsell_0.5.0          </span></span>
<span><span class="co">##  [79] scales_1.2.1            xtable_1.8-4            glue_1.6.2             </span></span>
<span><span class="co">##  [82] slam_0.1-50             tools_4.3.1             dendextend_1.17.1      </span></span>
<span><span class="co">##  [85] RSpectra_0.16-1         annotate_1.78.0         fs_1.6.3               </span></span>
<span><span class="co">##  [88] XML_3.99-0.14           Cairo_1.6-2             impute_1.74.1          </span></span>
<span><span class="co">##  [91] umap_0.2.10.0           AnnotationDbi_1.62.2    colorspace_2.1-0       </span></span>
<span><span class="co">##  [94] GenomeInfoDbData_1.2.10 cli_3.6.2               brew_1.0-8             </span></span>
<span><span class="co">##  [97] textshaping_0.3.7       fansi_1.0.5             eulerr_7.0.0           </span></span>
<span><span class="co">## [100] viridisLite_0.4.2       dplyr_1.1.3             gtable_0.3.4           </span></span>
<span><span class="co">## [103] sass_0.4.8              digest_0.6.33           BiocGenerics_0.48.1    </span></span>
<span><span class="co">## [106] rjson_0.2.21            memoise_2.0.1           htmltools_0.5.7        </span></span>
<span><span class="co">## [109] pkgdown_2.0.7           lifecycle_1.0.4         httr_1.4.7             </span></span>
<span><span class="co">## [112] GlobalOptions_0.1.2     openssl_2.1.1           microbenchmark_1.4.10  </span></span>
<span><span class="co">## [115] bit64_4.0.5</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Zuguang Gu.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  <style>nav[data-toggle='toc'] .nav .nav {display: block;}</style>
</body>
</html>
