<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Work with Big Datasets • cola</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Work with Big Datasets">
<meta property="og:description" content="cola">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">cola</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.9.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Standard consensus partitioning</li>
    <li>
      <a href="../articles/cola_quick.html">A Quick Start of Using cola Package</a>
    </li>
    <li>
      <a href="../articles/cola_general.html">A Framework for Consensus Partitioning</a>
    </li>
    <li>
      <a href="../articles/functional_enrichment.html">Automatic Functional Enrichment on Signature genes</a>
    </li>
    <li>
      <a href="../articles/predict.html">Predict Classes for New Samples</a>
    </li>
    <li>
      <a href="../articles/work_with_big_datasets.html">Work with Big Datasets</a>
    </li>
    <li>
      <a href="../articles/compare_partitions.html">Compare Partitions</a>
    </li>
    <li>
      <a href="../articles/ATC_methods.html">ATC - More Forms</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Hierarchical partitioning</li>
    <li>
      <a href="../articles/hierarchical.html">Hierarchical consensus partitioning</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Other examples</li>
    <li>
      <a href="../articles/AML.html">Re-analyze an AML proteomics dataset</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">On public datasets</li>
    <li>
      <a href="../articles/examples.html">Applications on public datasets</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Supplementary files</li>
    <li>
      <a href="https://jokergoo.github.io/cola_supplementary/" class="external-link">Supplementary files for cola NAR paper</a>
    </li>
    <li>
      <a href="https://cola-rh.github.io/supplementary/" class="external-link">Supplementary files for cola BIB paper</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/jokergoo/cola/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Work with Big Datasets</h1>
                        <h4 data-toc-skip class="author">Zuguang Gu (<a href="mailto:z.gu@dkfz.de" class="email">z.gu@dkfz.de</a>)</h4>
            
            <h4 data-toc-skip class="date">2024-02-27</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/jokergoo/cola/blob/HEAD/vignettes/work_with_big_datasets.Rmd" class="external-link"><code>vignettes/work_with_big_datasets.Rmd</code></a></small>
      <div class="hidden name"><code>work_with_big_datasets.Rmd</code></div>

    </div>

    
    
<p><em>cola</em> can be idealy applied to datasets with intermediate number of samples (around 500), however, the running time and the memory usage might not be acceptable if the number of samples gets very huge, e.g., more than thousands. Nevertheless, we can propose the following two strategies to partially solve this problem.</p>
<ol style="list-style-type: decimal">
<li>A randomly sampled subset of samples which take relatively short running time (e.g., 100-200 samples) can be firstly applied with cola, from which a specific combination of top-value method and partitioning method that gives the best results can be pre-selected. Later user can only apply these two specific methods to the complete dataset. This would be much faster than blindly running cola with many methods in sequence.</li>
<li>The random subset of samples can be treated as a training set to generate a classification of cells. Then, the class labels of the remaining samples can be predicted, e.g. by testing the distance to the centroid of each cell group, without rerunning consensus partitioning on them. cola implements a <code><a href="../reference/predict_classes-dispatch.html">predict_classes()</a></code> function for this purpose (see the vignette <a href="predict.html">“Predict classes for new samples”</a> for details).</li>
</ol>
<p>Note, since these two strategies are performed by sampling a small subset of cells from the cohort, the cell clusters with small size might not be detectable.</p>
<p>In the following examples, we use pseudo code to demonstrate the ideas. Assuming the full matrix is in an object <code>mat</code>. We randomly sample 200 samples from it:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ind</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/Bimap-envirAPI.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="../reference/ncol-dispatch.html">ncol</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span>, <span class="fl">200</span><span class="op">)</span></span>
<span><span class="va">mat1</span> <span class="op">=</span> <span class="va">mat</span><span class="op">[</span>, <span class="va">ind</span><span class="op">]</span></span></code></pre></div>
<p><strong>Strategy 1</strong>. First we can apply cola on <code>mat1</code>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rl</span> <span class="op">=</span> <span class="fu"><a href="../reference/run_all_consensus_partition_methods.html">run_all_consensus_partition_methods</a></span><span class="op">(</span><span class="va">mat1</span>, <span class="va">...</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/cola_report-dispatch.html">cola_report</a></span><span class="op">(</span><span class="va">rl</span>, <span class="va">...</span><span class="op">)</span></span></code></pre></div>
<p>Then find a best combination of top-value method and partitioining method. Assuming they are saved in objects <code>tm</code> and <code>pm</code>. Next run <code><a href="../reference/consensus_partition.html">consensus_partition()</a></code> on <code>mat</code> only with <code>tm</code> and <code>pm</code>:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">=</span> <span class="fu"><a href="../reference/consensus_partition.html">consensus_partition</a></span><span class="op">(</span><span class="va">mat</span>, top_value_method <span class="op">=</span> <span class="va">tm</span>, partition_method <span class="op">=</span> <span class="va">pm</span>, <span class="va">...</span><span class="op">)</span></span></code></pre></div>
<p><strong>Strategy 2</strong>. Similar as Strategy 1, we get the <code>ConsensusPartition</code> object from methods <code>tm</code> and <code>pm</code>, which was applied on <code>mat1</code>:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">=</span> <span class="va">rl</span><span class="op">[</span><span class="va">tm</span>, <span class="va">pm</span><span class="op">]</span></span></code></pre></div>
<p>Note the consensus partition object <code>res</code> is only based on a subset of original samples. With <code>res</code>, the classes of remaining samples can be predicted:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mat2</span> <span class="op">=</span> <span class="va">mat</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sets.html" class="external-link">setdiff</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="../reference/ncol-dispatch.html">ncol</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span><span class="op">)</span>, <span class="va">ind</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">mat2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/scale.html" class="external-link">scale</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">mat2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">cl</span> <span class="op">=</span> <span class="fu"><a href="../reference/predict_classes-dispatch.html">predict_classes</a></span><span class="op">(</span><span class="va">res</span>, <span class="va">mat2</span><span class="op">)</span></span></code></pre></div>
<p>Please note, by default <code>mat1</code> is scaled in cola analysis, correspondingly, <code>mat2</code> should also be row-scaled.</p>
<p>You can also directly send <code>mat</code> to <code><a href="../reference/predict_classes-dispatch.html">predict_classes()</a></code> function:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cl</span> <span class="op">=</span> <span class="fu"><a href="../reference/predict_classes-dispatch.html">predict_classes</a></span><span class="op">(</span><span class="va">res</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/scale.html" class="external-link">scale</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="the-downsamplingconsensuspartition-class">The <code>DownSamplingConsensusPartition</code> class<a class="anchor" aria-label="anchor" href="#the-downsamplingconsensuspartition-class"></a>
</h2>
<p><em>cola</em> implements a new <code>DownSamplingConsensusPartition</code> class for <strong>Strategy 2</strong> mentioned above. It runs consensus partitioning only on a subset of samples and the classes of other samples are predicted by <code><a href="../reference/predict_classes-dispatch.html">predict_classes()</a></code> internally. The <code>DownSamplingConsensusPartition</code> class inherits the <code>ConsensusPartition</code> class, so the functions that can be applied to the <code>ConsensusPartition</code> objects can also be applied to the <code>DownSamplingConsensusPartition</code> objects, just with some tiny difference.</p>
<p>In the following example, we demonstrate the usage of <code>DownSamplingConsensusPartition</code> class with the Golub dataset. Here the Golub dataset was used to generate another data object <code>golub_cola</code> and we can extract the matrix and the annotations by <code><a href="../reference/get_matrix-dispatch.html">get_matrix()</a></code> and <code><a href="../reference/get_anno-dispatch.html">get_anno()</a></code> functions.</p>
<p>To perform down sampling consensus partitioning, use the helper function <code><a href="../reference/consensus_partition_by_down_sampling.html">consensus_partition_by_down_sampling()</a></code>. This function basically run <code><a href="../reference/consensus_partition.html">consensus_partition()</a></code> on a subset of samples and later predict the classes for all samples by <code><a href="../reference/predict_classes-dispatch.html">predict_classes()</a></code> function. Here we set <code>subset</code> argument to 50, which means to randomly sample 50 samples from the whole dataset.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">golub_cola</span><span class="op">)</span></span>
<span><span class="va">m</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_matrix-dispatch.html">get_matrix</a></span><span class="op">(</span><span class="va">golub_cola</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">golub_cola_ds</span> <span class="op">=</span> <span class="fu"><a href="../reference/consensus_partition_by_down_sampling.html">consensus_partition_by_down_sampling</a></span><span class="op">(</span><span class="va">m</span>, subset <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  anno <span class="op">=</span> <span class="fu"><a href="../reference/get_anno-dispatch.html">get_anno</a></span><span class="op">(</span><span class="va">golub_cola</span><span class="op">)</span>, anno_col <span class="op">=</span> <span class="fu"><a href="../reference/get_anno_col-dispatch.html">get_anno_col</a></span><span class="op">(</span><span class="va">golub_cola</span><span class="op">)</span>,</span>
<span>  top_value_method <span class="op">=</span> <span class="st">"SD"</span>, partition_method <span class="op">=</span> <span class="st">"kmeans"</span><span class="op">)</span></span></code></pre></div>
<p>The object <code>golub_cola_ds</code> is already shipped in the package. Simply load the data object.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">golub_cola_ds</span><span class="op">)</span></span>
<span><span class="va">golub_cola_ds</span></span></code></pre></div>
<pre><code><span><span class="co">## A 'DownSamplingConsensusPartition' object with k = 2, 3, 4, 5, 6.</span></span>
<span><span class="co">##   On a matrix with 4116 rows and 50 columns, randomly sampled from 72 columns.</span></span>
<span><span class="co">##   Top rows (412) are extracted by 'ATC' method.</span></span>
<span><span class="co">##   Subgroups are detected by 'skmeans' method.</span></span>
<span><span class="co">##   Performed in total 250 partitions by row resampling.</span></span>
<span><span class="co">##   Best k for subgroups seems to be 2.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Following methods can be applied to this 'DownSamplingConsensusPartition' object:</span></span>
<span><span class="co">##  [1] "cola_report"             "collect_classes"         "collect_plots"          </span></span>
<span><span class="co">##  [4] "collect_stats"           "colnames"                "compare_partitions"     </span></span>
<span><span class="co">##  [7] "compare_signatures"      "consensus_heatmap"       "dimension_reduction"    </span></span>
<span><span class="co">## [10] "functional_enrichment"   "get_anno"                "get_anno_col"           </span></span>
<span><span class="co">## [13] "get_classes"             "get_consensus"           "get_matrix"             </span></span>
<span><span class="co">## [16] "get_membership"          "get_param"               "get_signatures"         </span></span>
<span><span class="co">## [19] "get_stats"               "is_best_k"               "is_stable_k"            </span></span>
<span><span class="co">## [22] "membership_heatmap"      "ncol"                    "nrow"                   </span></span>
<span><span class="co">## [25] "plot_ecdf"               "predict_classes"         "rownames"               </span></span>
<span><span class="co">## [28] "select_partition_number" "show"                    "suggest_best_k"         </span></span>
<span><span class="co">## [31] "test_to_known_factors"   "top_rows_heatmap"</span></span></code></pre>
<p>The summary of the <code>golub_cola_ds</code> is very similar as the <code>ConsensusPartition</code> object, except that it mentions the object is generated from 50 samples randomly sampled from 72 samples.</p>
<p>All the functions that are applied to the <code>ConsensusParition</code> class can be applied to the <code>DownSamplingConsensusPartition</code> class, except some tiny differences.</p>
<p><code><a href="../reference/get_classes-dispatch.html">get_classes()</a></code> returns the predicted classes for <em>all</em> samples:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">class</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_classes-dispatch.html">get_classes</a></span><span class="op">(</span><span class="va">golub_cola_ds</span>, k <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/nrow-dispatch.html">nrow</a></span><span class="op">(</span><span class="va">class</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 72</span></span></code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">class</span></span></code></pre></div>
<pre><code><span><span class="co">##           class     p</span></span>
<span><span class="co">## sample_39     2 0.000</span></span>
<span><span class="co">## sample_40     2 0.000</span></span>
<span><span class="co">## sample_42     1 0.000</span></span>
<span><span class="co">## sample_47     1 0.000</span></span>
<span><span class="co">## sample_48     1 0.000</span></span>
<span><span class="co">## sample_49     2 0.000</span></span>
<span><span class="co">## sample_41     1 0.000</span></span>
<span><span class="co">## sample_43     1 0.000</span></span>
<span><span class="co">## sample_44     1 0.000</span></span>
<span><span class="co">## sample_45     1 0.000</span></span>
<span><span class="co">## sample_46     1 0.000</span></span>
<span><span class="co">## sample_70     1 0.000</span></span>
<span><span class="co">## sample_71     1 0.000</span></span>
<span><span class="co">## sample_72     1 0.000</span></span>
<span><span class="co">## sample_68     1 0.000</span></span>
<span><span class="co">## sample_69     1 0.000</span></span>
<span><span class="co">## sample_67     1 0.000</span></span>
<span><span class="co">## sample_55     2 0.000</span></span>
<span><span class="co">## sample_56     2 0.000</span></span>
<span><span class="co">## sample_59     1 0.000</span></span>
<span><span class="co">## sample_52     2 0.000</span></span>
<span><span class="co">## sample_53     2 0.000</span></span>
<span><span class="co">## sample_51     2 0.000</span></span>
<span><span class="co">## sample_50     2 0.000</span></span>
<span><span class="co">## sample_54     1 0.000</span></span>
<span><span class="co">## sample_57     2 0.000</span></span>
<span><span class="co">## sample_58     2 0.000</span></span>
<span><span class="co">## sample_60     2 0.664</span></span>
<span><span class="co">## sample_61     2 0.000</span></span>
<span><span class="co">## sample_65     2 0.000</span></span>
<span><span class="co">## sample_66     1 0.000</span></span>
<span><span class="co">## sample_63     2 0.000</span></span>
<span><span class="co">## sample_64     2 0.000</span></span>
<span><span class="co">## sample_62     2 0.000</span></span>
<span><span class="co">## sample_1      2 0.000</span></span>
<span><span class="co">## sample_2      1 0.000</span></span>
<span><span class="co">## sample_3      2 0.000</span></span>
<span><span class="co">## sample_4      2 0.000</span></span>
<span><span class="co">## sample_5      1 0.000</span></span>
<span><span class="co">## sample_6      2 0.000</span></span>
<span><span class="co">## sample_7      2 0.000</span></span>
<span><span class="co">## sample_8      2 0.000</span></span>
<span><span class="co">## sample_9      1 0.000</span></span>
<span><span class="co">## sample_10     2 0.000</span></span>
<span><span class="co">## sample_11     1 0.000</span></span>
<span><span class="co">## sample_12     2 0.000</span></span>
<span><span class="co">## sample_13     1 0.000</span></span>
<span><span class="co">## sample_14     1 0.000</span></span>
<span><span class="co">## sample_15     1 0.000</span></span>
<span><span class="co">## sample_16     1 0.000</span></span>
<span><span class="co">## sample_17     1 0.000</span></span>
<span><span class="co">## sample_18     2 0.000</span></span>
<span><span class="co">## sample_19     1 0.000</span></span>
<span><span class="co">## sample_20     1 0.000</span></span>
<span><span class="co">## sample_21     1 0.000</span></span>
<span><span class="co">## sample_22     2 0.000</span></span>
<span><span class="co">## sample_23     2 0.000</span></span>
<span><span class="co">## sample_24     1 0.000</span></span>
<span><span class="co">## sample_25     2 0.000</span></span>
<span><span class="co">## sample_26     1 0.000</span></span>
<span><span class="co">## sample_27     2 0.000</span></span>
<span><span class="co">## sample_34     2 0.000</span></span>
<span><span class="co">## sample_35     2 0.000</span></span>
<span><span class="co">## sample_36     2 0.000</span></span>
<span><span class="co">## sample_37     2 0.000</span></span>
<span><span class="co">## sample_38     2 0.000</span></span>
<span><span class="co">## sample_28     2 0.000</span></span>
<span><span class="co">## sample_29     1 0.000</span></span>
<span><span class="co">## sample_30     2 0.000</span></span>
<span><span class="co">## sample_31     2 0.000</span></span>
<span><span class="co">## sample_32     2 0.000</span></span>
<span><span class="co">## sample_33     2 0.000</span></span></code></pre>
<p>There is an additional column named <code>p</code> which is the p-value for predicting the class labels. For more details of how the p-value is calculated, please refer to the vignette <a href="predict.html">“Predict classes for new samples”</a>.</p>
<p>If the argument <code>k</code> is not specified or <code>k</code> is specified as a vector, the class labels for all <code>k</code> are returned. Now you can set the <code>p_cutoff</code> argument so that the class label with p-value larger than this is set to <code>NA</code>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/get_classes-dispatch.html">get_classes</a></span><span class="op">(</span><span class="va">golub_cola_ds</span>, p_cutoff <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##           k=2 k=3 k=4 k=5 k=6</span></span>
<span><span class="co">## sample_39   2   2   2   2  NA</span></span>
<span><span class="co">## sample_40   2   3   3   3  NA</span></span>
<span><span class="co">## sample_42   1   1   4   4   3</span></span>
<span><span class="co">## sample_47   1   1   1   1   1</span></span>
<span><span class="co">## sample_48   1   1   1   1   1</span></span>
<span><span class="co">## sample_49   2   3   3  NA   2</span></span>
<span><span class="co">## sample_41   1   1   1   1   1</span></span>
<span><span class="co">## sample_43   1   1   1   1   1</span></span>
<span><span class="co">## sample_44   1   1   1   1   1</span></span>
<span><span class="co">## sample_45   1   1   1   1   1</span></span>
<span><span class="co">## sample_46   1   1   1  NA  NA</span></span>
<span><span class="co">## sample_70   1  NA  NA  NA  NA</span></span>
<span><span class="co">## sample_71   1  NA   4   4   3</span></span>
<span><span class="co">## sample_72   1   1   4   4   3</span></span>
<span><span class="co">## sample_68   1   1   1   1   1</span></span>
<span><span class="co">## sample_69   1   1   1   1   1</span></span>
<span><span class="co">## sample_67   1   1   4   4   3</span></span>
<span><span class="co">## sample_55   2   3   3  NA  NA</span></span>
<span><span class="co">## sample_56   2   3   3  NA  NA</span></span>
<span><span class="co">## sample_59   1   1  NA  NA  NA</span></span>
<span><span class="co">## sample_52   2   2   2  NA  NA</span></span>
<span><span class="co">## sample_53   2   2   2  NA  NA</span></span>
<span><span class="co">## sample_51   2   2   2   2   2</span></span>
<span><span class="co">## sample_50   2   2   2   2   2</span></span>
<span><span class="co">## sample_54   1   1   1  NA  NA</span></span>
<span><span class="co">## sample_57   2   2   2   2   2</span></span>
<span><span class="co">## sample_58   2   2   2   2   2</span></span>
<span><span class="co">## sample_60  NA   1  NA  NA  NA</span></span>
<span><span class="co">## sample_61   2   2   2   2   2</span></span>
<span><span class="co">## sample_65   2   2   2   2   2</span></span>
<span><span class="co">## sample_66   1   1   1   1   1</span></span>
<span><span class="co">## sample_63   2   2   2   2   2</span></span>
<span><span class="co">## sample_64   2   2   2   2   2</span></span>
<span><span class="co">## sample_62   2   2   2   2   2</span></span>
<span><span class="co">## sample_1    2   3   3   3  NA</span></span>
<span><span class="co">## sample_2    1   1   4   4   3</span></span>
<span><span class="co">## sample_3    2   3   3   3   3</span></span>
<span><span class="co">## sample_4    2   3   3   3  NA</span></span>
<span><span class="co">## sample_5    1   1   1   1   1</span></span>
<span><span class="co">## sample_6    2   3   3   3  NA</span></span>
<span><span class="co">## sample_7    2   3   3   3   2</span></span>
<span><span class="co">## sample_8    2   3   3  NA   2</span></span>
<span><span class="co">## sample_9    1   1   1  NA  NA</span></span>
<span><span class="co">## sample_10   2   3   3  NA   3</span></span>
<span><span class="co">## sample_11   1   1   4   4   3</span></span>
<span><span class="co">## sample_12   2   2   2   2   2</span></span>
<span><span class="co">## sample_13   1   1   1   1   1</span></span>
<span><span class="co">## sample_14   1   1   1   1  NA</span></span>
<span><span class="co">## sample_15   1   1   1   1   1</span></span>
<span><span class="co">## sample_16   1   1   1   1   1</span></span>
<span><span class="co">## sample_17   1   1   1   1  NA</span></span>
<span><span class="co">## sample_18   2   3   3  NA  NA</span></span>
<span><span class="co">## sample_19   1   1   1   1   1</span></span>
<span><span class="co">## sample_20   1   1   1   1   1</span></span>
<span><span class="co">## sample_21   1   1   1   1   1</span></span>
<span><span class="co">## sample_22   2   2  NA  NA  NA</span></span>
<span><span class="co">## sample_23   2   3   3   3   3</span></span>
<span><span class="co">## sample_24   1   1   1   1   1</span></span>
<span><span class="co">## sample_25   2   2   2  NA  NA</span></span>
<span><span class="co">## sample_26   1   1  NA  NA  NA</span></span>
<span><span class="co">## sample_27   2   3   3   3   2</span></span>
<span><span class="co">## sample_34   2   2   2   2   2</span></span>
<span><span class="co">## sample_35   2   2   2   2   2</span></span>
<span><span class="co">## sample_36   2   2   2   2   2</span></span>
<span><span class="co">## sample_37   2   2   2   2   2</span></span>
<span><span class="co">## sample_38   2   2   2   2   2</span></span>
<span><span class="co">## sample_28   2   2   2   2  NA</span></span>
<span><span class="co">## sample_29   1   1   4   4   3</span></span>
<span><span class="co">## sample_30   2   2   2  NA  NA</span></span>
<span><span class="co">## sample_31   2   2   2   2   2</span></span>
<span><span class="co">## sample_32   2   2   2   2   2</span></span>
<span><span class="co">## sample_33   2   2   2   2   2</span></span></code></pre>
<p>There are several functions uses <code>p_cutoff</code> argument which controls the number of “usable samples” or the samples with reliable classifications. These functions are <code><a href="../reference/get_classes-dispatch.html">get_classes()</a></code>, <code><a href="../reference/test_to_known_factors-dispatch.html">test_to_known_factors()</a></code>, <code><a href="../reference/dimension_reduction-dispatch.html">dimension_reduction()</a></code> and <code><a href="../reference/get_signatures-dispatch.html">get_signatures()</a></code>.</p>
<p>For <code><a href="../reference/dimension_reduction-dispatch.html">dimension_reduction()</a></code> function, the samples with p-value higher than <code>p_cutoff</code> are marked by crosses. The samples that were not selected in the random sampling were mapped with smaller dots.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/dimension_reduction-dispatch.html">dimension_reduction</a></span><span class="op">(</span><span class="va">golub_cola_ds</span>, k <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## use UMAP</span></span></code></pre>
<p><img src="work_with_big_datasets_files/figure-html/unnamed-chunk-12-1.png" width="500" style="display: block; margin: auto;"></p>
<p>For <code><a href="../reference/get_signatures-dispatch.html">get_signatures()</a></code> function, the signatures are only found in the samples with p-values less than <code>p_cutoff</code>.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/get_signatures-dispatch.html">get_signatures</a></span><span class="op">(</span><span class="va">golub_cola_ds</span>, k <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## * 72/72 samples (in 2 classes) remain after filtering by p-value (&lt;= 1).</span></span>
<span><span class="co">## * cache hash: f69900f48bd1fbe0f5185b1b5ab034a0 (seed 888).</span></span>
<span><span class="co">## * calculating row difference between subgroups by Ftest.</span></span>
<span><span class="co">## * split rows into 2 groups by k-means clustering.</span></span>
<span><span class="co">## * 1802 signatures (43.8%) under fdr &lt; 0.05, group_diff &gt; 0.</span></span>
<span><span class="co">## * making heatmaps for signatures.</span></span></code></pre>
<p><img src="work_with_big_datasets_files/figure-html/unnamed-chunk-13-1.png" width="500" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">## Platform: x86_64-apple-darwin20 (64-bit)</span></span>
<span><span class="co">## Running under: macOS Ventura 13.2.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Random number generation:</span></span>
<span><span class="co">##  RNG:     L'Ecuyer-CMRG </span></span>
<span><span class="co">##  Normal:  Inversion </span></span>
<span><span class="co">##  Sample:  Rejection </span></span>
<span><span class="co">##  </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] C/UTF-8/C/C/C/C</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Europe/Berlin</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] cola_2.9.1    knitr_1.44    markdown_1.10</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] blob_1.2.4              bitops_1.0-7            Biostrings_2.68.1       RCurl_1.98-1.12        </span></span>
<span><span class="co">##  [5] fastmap_1.1.1           XML_3.99-0.14           digest_0.6.33           lifecycle_1.0.4        </span></span>
<span><span class="co">##  [9] cluster_2.1.4           Cairo_1.6-2             survival_3.5-8          KEGGREST_1.40.1        </span></span>
<span><span class="co">## [13] RSQLite_2.3.1           magrittr_2.0.3          genefilter_1.82.1       compiler_4.3.1         </span></span>
<span><span class="co">## [17] rlang_1.1.2             sass_0.4.8              rngtools_1.5.2          tools_4.3.1            </span></span>
<span><span class="co">## [21] yaml_2.3.7              askpass_1.2.0           brew_1.0-8              doRNG_1.8.6            </span></span>
<span><span class="co">## [25] bit_4.0.5               mclust_6.0.0            reticulate_1.34.0       xml2_1.3.6             </span></span>
<span><span class="co">## [29] eulerr_7.0.0            RColorBrewer_1.1-3      purrr_1.0.2             BiocGenerics_0.48.1    </span></span>
<span><span class="co">## [33] desc_1.4.2              grid_4.3.1              stats4_4.3.1            xtable_1.8-4           </span></span>
<span><span class="co">## [37] colorspace_2.1-0        iterators_1.0.14        cli_3.6.2               rmarkdown_2.25         </span></span>
<span><span class="co">## [41] crayon_1.5.2            ragg_1.2.6              umap_0.2.10.0           RSpectra_0.16-1        </span></span>
<span><span class="co">## [45] httr_1.4.7              rjson_0.2.21            DBI_1.1.3               cachem_1.0.8           </span></span>
<span><span class="co">## [49] stringr_1.5.0           splines_4.3.1           zlibbioc_1.46.0         parallel_4.3.1         </span></span>
<span><span class="co">## [53] AnnotationDbi_1.62.2    impute_1.74.1           XVector_0.40.0          matrixStats_1.2.0      </span></span>
<span><span class="co">## [57] vctrs_0.6.4             Matrix_1.6-1.1          jsonlite_1.8.8          slam_0.1-50            </span></span>
<span><span class="co">## [61] IRanges_2.36.0          GetoptLong_1.0.5        S4Vectors_0.40.2        bit64_4.0.5            </span></span>
<span><span class="co">## [65] irlba_2.3.5.1           clue_0.3-65             magick_2.8.0            systemfonts_1.0.5      </span></span>
<span><span class="co">## [69] foreach_1.5.2           jquerylib_0.1.4         annotate_1.78.0         glue_1.6.2             </span></span>
<span><span class="co">## [73] pkgdown_2.0.7           codetools_0.2-19        stringi_1.7.12          shape_1.4.6            </span></span>
<span><span class="co">## [77] GenomeInfoDb_1.36.4     ComplexHeatmap_2.18.0   htmltools_0.5.7         openssl_2.1.1          </span></span>
<span><span class="co">## [81] GenomeInfoDbData_1.2.10 circlize_0.4.15         R6_2.5.1                microbenchmark_1.4.10  </span></span>
<span><span class="co">## [85] textshaping_0.3.7       doParallel_1.0.17       rprojroot_2.0.3         evaluate_0.22          </span></span>
<span><span class="co">## [89] lattice_0.21-9          Biobase_2.60.0          png_0.1-8               memoise_2.0.1          </span></span>
<span><span class="co">## [93] bslib_0.6.1             Rcpp_1.0.11             xfun_0.40               fs_1.6.3               </span></span>
<span><span class="co">## [97] MatrixGenerics_1.12.3   skmeans_0.2-16          GlobalOptions_0.1.2</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Zuguang Gu.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  <style>nav[data-toggle='toc'] .nav .nav {display: block;}</style>
</body>
</html>
